{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use annot conda environment\n",
    "# See https://medium.com/intothegenomics/annotate-genes-and-genomic-coordinates-using-python-9259efa6ffc2\n",
    "\n",
    "#https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "\n",
    "annotation_file = \"/Users/meghansleeper/Desktop/farm-files/dmr-workflow/supplementary/annotation_beds/gencode.v44.annotation.all.txt.sorted.formatted.bed.gz\"\n",
    "DMR_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers.tsv\"\n",
    "output_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv\"\n",
    "\n",
    "gencode_v44 = pysam.TabixFile(annotation_file)\n",
    "\n",
    "# Annotation file columns: ['seqname',\n",
    "                        #  'start',\n",
    "                        #  'end',\n",
    "                        #  'attribute',\n",
    "                        #  'feature',\n",
    "                        #  'strand',\n",
    "                        #  'source',\n",
    "                        #  'gene_name',\n",
    "                        #  'gene_id',\n",
    "                        #  'gene_type',\n",
    "                        #  'gene_level',\n",
    "                        #  'exon_id',\n",
    "                        #  'exon_number',\n",
    "                        #  'transcript_name',\n",
    "                        #  'transcript_type']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>56140</td>\n",
       "      <td>58080</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>ENSG00000278704</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003753029.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000618686</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>56140</td>\n",
       "      <td>58376</td>\n",
       "      <td>exon</td>\n",
       "      <td>ENSG00000278704</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003753029.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000618686</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>56140</td>\n",
       "      <td>58376</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000278704</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>56140</td>\n",
       "      <td>58376</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000278704</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000618686</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>58081</td>\n",
       "      <td>58083</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>ENSG00000278704</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003753029.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000618686</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  start    end          feature        gene_name       gene_type  \\\n",
       "0  GL000009.2  56140  58080  three_prime_UTR  ENSG00000278704  protein_coding   \n",
       "1  GL000009.2  56140  58376             exon  ENSG00000278704  protein_coding   \n",
       "2  GL000009.2  56140  58376             gene  ENSG00000278704  protein_coding   \n",
       "3  GL000009.2  56140  58376       transcript  ENSG00000278704  protein_coding   \n",
       "4  GL000009.2  58081  58083       stop_codon  ENSG00000278704  protein_coding   \n",
       "\n",
       "   gene_level            exon_id  exon_number  transcript_name transcript_type  \n",
       "0           3  ENSE00003753029.1          1.0  ENST00000618686  protein_coding  \n",
       "1           3  ENSE00003753029.1          1.0  ENST00000618686  protein_coding  \n",
       "2           3                NaN          NaN              NaN             NaN  \n",
       "3           3                NaN          NaN  ENST00000618686  protein_coding  \n",
       "4           3  ENSE00003753029.1          1.0  ENST00000618686  protein_coding  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df = pd.read_csv(annotation_file, sep='\\t', header=None, names=['chr', 'start', 'end', 'feature', 'gene_name', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type'])\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2192410</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>transcript</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L16-207</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192411</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213204</td>\n",
       "      <td>57213357</td>\n",
       "      <td>exon</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015126.1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>DDX11L16-207</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192412</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213526</td>\n",
       "      <td>57213602</td>\n",
       "      <td>exon</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015125.1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>DDX11L16-207</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192413</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213880</td>\n",
       "      <td>57213964</td>\n",
       "      <td>exon</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015124.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>DDX11L16-207</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192414</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214350</td>\n",
       "      <td>57214397</td>\n",
       "      <td>exon</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015123.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>DDX11L16-207</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     start       end     feature gene_name  \\\n",
       "2192410  chrY  57212184  57214397  transcript  DDX11L16   \n",
       "2192411  chrY  57213204  57213357        exon  DDX11L16   \n",
       "2192412  chrY  57213526  57213602        exon  DDX11L16   \n",
       "2192413  chrY  57213880  57213964        exon  DDX11L16   \n",
       "2192414  chrY  57214350  57214397        exon  DDX11L16   \n",
       "\n",
       "                      gene_type  gene_level            exon_id  exon_number  \\\n",
       "2192410  unprocessed_pseudogene           2                NaN          NaN   \n",
       "2192411  unprocessed_pseudogene           2  ENSE00004015126.1          4.0   \n",
       "2192412  unprocessed_pseudogene           2  ENSE00004015125.1          3.0   \n",
       "2192413  unprocessed_pseudogene           2  ENSE00004015124.1          2.0   \n",
       "2192414  unprocessed_pseudogene           2  ENSE00004015123.1          1.0   \n",
       "\n",
       "        transcript_name         transcript_type  \n",
       "2192410    DDX11L16-207  unprocessed_pseudogene  \n",
       "2192411    DDX11L16-207  unprocessed_pseudogene  \n",
       "2192412    DDX11L16-207  unprocessed_pseudogene  \n",
       "2192413    DDX11L16-207  unprocessed_pseudogene  \n",
       "2192414    DDX11L16-207  unprocessed_pseudogene  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>186803</th>\n",
       "      <td>MU273376.1</td>\n",
       "      <td>6960</td>\n",
       "      <td>6962</td>\n",
       "      <td>stop_codon_redefined_as_selenocysteine</td>\n",
       "      <td>MSRB1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSRB1-206</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227362</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25802093</td>\n",
       "      <td>25802095</td>\n",
       "      <td>stop_codon_redefined_as_selenocysteine</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SELENON-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227406</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25812789</td>\n",
       "      <td>25812791</td>\n",
       "      <td>stop_codon_redefined_as_selenocysteine</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SELENON-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227407</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25812789</td>\n",
       "      <td>25812791</td>\n",
       "      <td>stop_codon_redefined_as_selenocysteine</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SELENON-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227408</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25812789</td>\n",
       "      <td>25812791</td>\n",
       "      <td>stop_codon_redefined_as_selenocysteine</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SELENON-203</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               chr     start       end  \\\n",
       "186803  MU273376.1      6960      6962   \n",
       "227362        chr1  25802093  25802095   \n",
       "227406        chr1  25812789  25812791   \n",
       "227407        chr1  25812789  25812791   \n",
       "227408        chr1  25812789  25812791   \n",
       "\n",
       "                                       feature gene_name       gene_type  \\\n",
       "186803  stop_codon_redefined_as_selenocysteine     MSRB1  protein_coding   \n",
       "227362  stop_codon_redefined_as_selenocysteine   SELENON  protein_coding   \n",
       "227406  stop_codon_redefined_as_selenocysteine   SELENON  protein_coding   \n",
       "227407  stop_codon_redefined_as_selenocysteine   SELENON  protein_coding   \n",
       "227408  stop_codon_redefined_as_selenocysteine   SELENON  protein_coding   \n",
       "\n",
       "        gene_level exon_id  exon_number transcript_name transcript_type  \n",
       "186803           2     NaN          NaN       MSRB1-206  protein_coding  \n",
       "227362           2     NaN          NaN     SELENON-202  protein_coding  \n",
       "227406           2     NaN          NaN     SELENON-201  protein_coding  \n",
       "227407           2     NaN          NaN     SELENON-202  protein_coding  \n",
       "227408           2     NaN          NaN     SELENON-203  protein_coding  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df[annot_df.feature=='stop_codon_redefined_as_selenocysteine'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GL000009.2', 'GL000194.1', 'GL000195.1', 'GL000205.2',\n",
       "       'GL000209.2', 'GL000213.1', 'GL000216.2', 'GL000218.1',\n",
       "       'GL000219.1', 'GL000220.1', 'GL000225.1', 'GL000250.2',\n",
       "       'GL000251.2', 'GL000252.2', 'GL000253.2', 'GL000254.2',\n",
       "       'GL000255.2', 'GL000256.2', 'GL000257.2', 'GL000258.2',\n",
       "       'GL339449.2', 'GL383518.1', 'GL383519.1', 'GL383520.2',\n",
       "       'GL383521.1', 'GL383522.1', 'GL383526.1', 'GL383527.1',\n",
       "       'GL383528.1', 'GL383531.1', 'GL383532.1', 'GL383533.1',\n",
       "       'GL383534.2', 'GL383539.1', 'GL383540.1', 'GL383541.1',\n",
       "       'GL383542.1', 'GL383545.1', 'GL383546.1', 'GL383550.2',\n",
       "       'GL383551.1', 'GL383552.1', 'GL383553.2', 'GL383554.1',\n",
       "       'GL383555.2', 'GL383556.1', 'GL383557.1', 'GL383563.3',\n",
       "       'GL383564.2', 'GL383565.1', 'GL383566.1', 'GL383567.1',\n",
       "       'GL383569.1', 'GL383570.1', 'GL383571.1', 'GL383572.1',\n",
       "       'GL383573.1', 'GL383574.1', 'GL383575.2', 'GL383576.1',\n",
       "       'GL383577.2', 'GL383579.2', 'GL383580.2', 'GL383581.2',\n",
       "       'GL383582.2', 'GL383583.2', 'GL582966.2', 'GL877875.1',\n",
       "       'GL877876.1', 'GL949742.1', 'GL949746.1', 'GL949747.2',\n",
       "       'GL949748.2', 'GL949749.2', 'GL949750.2', 'GL949751.2',\n",
       "       'GL949752.1', 'GL949753.2', 'JH159136.1', 'JH159137.1',\n",
       "       'JH159146.1', 'JH159147.1', 'JH159148.1', 'JH636055.2',\n",
       "       'KB021644.2', 'KB663609.1', 'KI270442.1', 'KI270711.1',\n",
       "       'KI270713.1', 'KI270721.1', 'KI270726.1', 'KI270727.1',\n",
       "       'KI270728.1', 'KI270731.1', 'KI270733.1', 'KI270734.1',\n",
       "       'KI270744.1', 'KI270750.1', 'KI270758.1', 'KI270759.1',\n",
       "       'KI270760.1', 'KI270761.1', 'KI270762.1', 'KI270763.1',\n",
       "       'KI270765.1', 'KI270766.1', 'KI270767.1', 'KI270768.1',\n",
       "       'KI270769.1', 'KI270770.1', 'KI270772.1', 'KI270773.1',\n",
       "       'KI270774.1', 'KI270776.1', 'KI270777.1', 'KI270779.1',\n",
       "       'KI270780.1', 'KI270781.1', 'KI270782.1', 'KI270783.1',\n",
       "       'KI270784.1', 'KI270785.1', 'KI270786.1', 'KI270788.1',\n",
       "       'KI270789.1', 'KI270790.1', 'KI270791.1', 'KI270792.1',\n",
       "       'KI270793.1', 'KI270794.1', 'KI270795.1', 'KI270796.1',\n",
       "       'KI270797.1', 'KI270798.1', 'KI270799.1', 'KI270800.1',\n",
       "       'KI270801.1', 'KI270802.1', 'KI270803.1', 'KI270804.1',\n",
       "       'KI270805.1', 'KI270806.1', 'KI270807.1', 'KI270808.1',\n",
       "       'KI270809.1', 'KI270810.1', 'KI270811.1', 'KI270812.1',\n",
       "       'KI270813.1', 'KI270814.1', 'KI270815.1', 'KI270816.1',\n",
       "       'KI270817.1', 'KI270818.1', 'KI270819.1', 'KI270821.1',\n",
       "       'KI270822.1', 'KI270823.1', 'KI270824.1', 'KI270825.1',\n",
       "       'KI270829.1', 'KI270830.1', 'KI270831.1', 'KI270832.1',\n",
       "       'KI270833.1', 'KI270834.1', 'KI270835.1', 'KI270837.1',\n",
       "       'KI270838.1', 'KI270840.1', 'KI270842.1', 'KI270844.1',\n",
       "       'KI270845.1', 'KI270846.1', 'KI270847.1', 'KI270848.1',\n",
       "       'KI270849.1', 'KI270850.1', 'KI270851.1', 'KI270852.1',\n",
       "       'KI270853.1', 'KI270854.1', 'KI270855.1', 'KI270856.1',\n",
       "       'KI270857.1', 'KI270858.1', 'KI270859.1', 'KI270860.1',\n",
       "       'KI270861.1', 'KI270862.1', 'KI270863.1', 'KI270865.1',\n",
       "       'KI270866.1', 'KI270867.1', 'KI270868.1', 'KI270869.1',\n",
       "       'KI270870.1', 'KI270871.1', 'KI270872.1', 'KI270873.1',\n",
       "       'KI270874.1', 'KI270875.1', 'KI270876.1', 'KI270877.1',\n",
       "       'KI270878.1', 'KI270879.1', 'KI270880.1', 'KI270881.1',\n",
       "       'KI270882.1', 'KI270883.1', 'KI270884.1', 'KI270885.1',\n",
       "       'KI270886.1', 'KI270887.1', 'KI270888.1', 'KI270889.1',\n",
       "       'KI270890.1', 'KI270891.1', 'KI270892.1', 'KI270893.1',\n",
       "       'KI270894.1', 'KI270895.1', 'KI270896.1', 'KI270897.1',\n",
       "       'KI270898.1', 'KI270899.1', 'KI270900.1', 'KI270901.1',\n",
       "       'KI270902.1', 'KI270903.1', 'KI270904.1', 'KI270905.1',\n",
       "       'KI270906.1', 'KI270907.1', 'KI270908.1', 'KI270909.1',\n",
       "       'KI270910.1', 'KI270911.1', 'KI270912.1', 'KI270913.1',\n",
       "       'KI270914.1', 'KI270915.1', 'KI270916.1', 'KI270917.1',\n",
       "       'KI270918.1', 'KI270919.1', 'KI270920.1', 'KI270921.1',\n",
       "       'KI270922.1', 'KI270923.1', 'KI270924.1', 'KI270925.1',\n",
       "       'KI270926.1', 'KI270927.1', 'KI270928.1', 'KI270929.1',\n",
       "       'KI270930.1', 'KI270931.1', 'KI270932.1', 'KI270933.1',\n",
       "       'KI270934.1', 'KI270935.1', 'KI270936.1', 'KI270937.1',\n",
       "       'KI270938.1', 'KN196472.1', 'KN196473.1', 'KN196474.1',\n",
       "       'KN196475.1', 'KN196476.1', 'KN196477.1', 'KN196478.1',\n",
       "       'KN196479.1', 'KN196480.1', 'KN196481.1', 'KN196484.1',\n",
       "       'KN196485.1', 'KN196486.1', 'KN196487.1', 'KN538360.1',\n",
       "       'KN538361.1', 'KN538362.1', 'KN538363.1', 'KN538364.1',\n",
       "       'KN538368.1', 'KN538369.1', 'KN538370.1', 'KN538371.1',\n",
       "       'KN538372.1', 'KN538373.1', 'KQ031383.1', 'KQ031384.1',\n",
       "       'KQ031385.1', 'KQ031387.1', 'KQ031388.1', 'KQ031389.1',\n",
       "       'KQ090014.1', 'KQ090015.1', 'KQ090016.1', 'KQ090018.1',\n",
       "       'KQ090021.1', 'KQ090022.1', 'KQ090026.1', 'KQ090027.1',\n",
       "       'KQ090028.1', 'KQ458382.1', 'KQ458383.1', 'KQ458384.1',\n",
       "       'KQ458385.1', 'KQ458386.1', 'KQ458387.1', 'KQ458388.1',\n",
       "       'KQ759759.2', 'KQ759760.1', 'KQ759761.1', 'KQ759762.2',\n",
       "       'KQ983255.1', 'KQ983256.1', 'KQ983257.1', 'KQ983258.1',\n",
       "       'KV575243.1', 'KV575244.1', 'KV575245.1', 'KV575246.1',\n",
       "       'KV575247.1', 'KV575248.1', 'KV575249.1', 'KV575250.1',\n",
       "       'KV575251.1', 'KV575252.1', 'KV575253.1', 'KV575254.1',\n",
       "       'KV575255.1', 'KV575256.1', 'KV575257.1', 'KV575258.1',\n",
       "       'KV575259.1', 'KV575260.1', 'KV766192.1', 'KV766193.1',\n",
       "       'KV766194.1', 'KV766195.1', 'KV766196.1', 'KV766198.1',\n",
       "       'KV880763.1', 'KV880764.1', 'KV880765.1', 'KV880766.1',\n",
       "       'KV880768.1', 'KZ208904.1', 'KZ208905.1', 'KZ208906.1',\n",
       "       'KZ208907.1', 'KZ208908.1', 'KZ208909.1', 'KZ208911.1',\n",
       "       'KZ208912.1', 'KZ208913.1', 'KZ208914.1', 'KZ208915.1',\n",
       "       'KZ208916.1', 'KZ208917.1', 'KZ208919.1', 'KZ208920.1',\n",
       "       'KZ208921.1', 'KZ208922.1', 'KZ208924.1', 'KZ559102.1',\n",
       "       'KZ559103.1', 'KZ559104.1', 'KZ559105.1', 'KZ559106.1',\n",
       "       'KZ559107.1', 'KZ559108.1', 'KZ559109.1', 'KZ559110.1',\n",
       "       'KZ559111.1', 'KZ559112.1', 'KZ559113.1', 'KZ559114.1',\n",
       "       'KZ559115.1', 'KZ559116.1', 'ML143341.1', 'ML143342.1',\n",
       "       'ML143343.1', 'ML143344.1', 'ML143345.1', 'ML143347.1',\n",
       "       'ML143349.1', 'ML143350.1', 'ML143352.1', 'ML143354.1',\n",
       "       'ML143355.1', 'ML143357.1', 'ML143358.1', 'ML143359.1',\n",
       "       'ML143360.1', 'ML143361.1', 'ML143362.1', 'ML143365.1',\n",
       "       'ML143366.1', 'ML143367.1', 'ML143369.1', 'ML143370.1',\n",
       "       'ML143371.1', 'ML143372.1', 'ML143373.1', 'ML143374.1',\n",
       "       'ML143375.1', 'ML143376.1', 'ML143377.1', 'ML143378.1',\n",
       "       'ML143380.1', 'ML143381.1', 'MU273330.1', 'MU273331.1',\n",
       "       'MU273332.1', 'MU273333.1', 'MU273334.1', 'MU273335.1',\n",
       "       'MU273336.1', 'MU273337.1', 'MU273338.1', 'MU273339.1',\n",
       "       'MU273340.1', 'MU273341.1', 'MU273342.1', 'MU273343.1',\n",
       "       'MU273344.1', 'MU273345.1', 'MU273346.1', 'MU273347.1',\n",
       "       'MU273348.1', 'MU273349.1', 'MU273351.1', 'MU273352.1',\n",
       "       'MU273353.1', 'MU273355.1', 'MU273356.1', 'MU273357.1',\n",
       "       'MU273358.1', 'MU273359.1', 'MU273360.1', 'MU273361.1',\n",
       "       'MU273362.1', 'MU273363.1', 'MU273364.1', 'MU273365.1',\n",
       "       'MU273366.1', 'MU273367.1', 'MU273368.1', 'MU273369.1',\n",
       "       'MU273370.1', 'MU273371.1', 'MU273372.1', 'MU273373.1',\n",
       "       'MU273374.1', 'MU273375.1', 'MU273376.1', 'MU273377.1',\n",
       "       'MU273378.1', 'MU273379.1', 'MU273380.1', 'MU273381.1',\n",
       "       'MU273382.1', 'MU273383.1', 'MU273384.1', 'MU273385.1',\n",
       "       'MU273387.1', 'MU273388.1', 'MU273389.1', 'MU273390.1',\n",
       "       'MU273391.1', 'MU273392.1', 'MU273393.1', 'MU273394.1',\n",
       "       'MU273395.1', 'MU273396.1', 'MU273397.1', 'MU273398.1', 'chr1',\n",
       "       'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16',\n",
       "       'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21', 'chr22',\n",
       "       'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrM',\n",
       "       'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# figure out how to interpret the chromosome labels\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=\n",
    "annot_df.chr.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>bg_mean</th>\n",
       "      <th>delta_means</th>\n",
       "      <th>delta_quants</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6CpGs</td>\n",
       "      <td>416bp</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.413</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7CpGs</td>\n",
       "      <td>468bp</td>\n",
       "      <td>0.260</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8CpGs</td>\n",
       "      <td>206bp</td>\n",
       "      <td>0.633</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.356</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11CpGs</td>\n",
       "      <td>1811bp</td>\n",
       "      <td>0.236</td>\n",
       "      <td>0.625</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22CpGs</td>\n",
       "      <td>1169bp</td>\n",
       "      <td>0.435</td>\n",
       "      <td>0.875</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "   lenCpG      bp  tg_mean  bg_mean  delta_means  delta_quants  delta_maxmin  \\\n",
       "0   6CpGs   416bp    0.311    0.724        0.413         0.372         0.330   \n",
       "1   7CpGs   468bp    0.260    0.667        0.407         0.360         0.314   \n",
       "2   8CpGs   206bp    0.633    1.000        0.367         0.356         0.344   \n",
       "3  11CpGs  1811bp    0.236    0.625        0.389         0.382         0.375   \n",
       "4  22CpGs  1169bp    0.435    0.875        0.440         0.407         0.375   \n",
       "\n",
       "    ttest direction  \n",
       "0  0.1260         U  \n",
       "1  0.1460         U  \n",
       "2  0.0761         U  \n",
       "3  0.1640         U  \n",
       "4  0.0927         U  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Read the sample file in to a pandas dataframe\n",
    "df = pd.read_table(DMR_file, low_memory=False, header=0)\n",
    "\n",
    "# figure out syntax to add data type to silence warning --> must be specified dtype like np.int16 \n",
    "# see: https://stackoverflow.com/questions/24251219/pandas-read-csv-low-memory-and-dtype-options\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>bg_mean</th>\n",
       "      <th>delta_means</th>\n",
       "      <th>delta_quants</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.413</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.356</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>0.625</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>0.875</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "  lenCpG    bp  tg_mean  bg_mean  delta_means  delta_quants  delta_maxmin  \\\n",
       "0      6   416    0.311    0.724        0.413         0.372         0.330   \n",
       "1      7   468    0.260    0.667        0.407         0.360         0.314   \n",
       "2      8   206    0.633    1.000        0.367         0.356         0.344   \n",
       "3     11  1811    0.236    0.625        0.389         0.382         0.375   \n",
       "4     22  1169    0.435    0.875        0.440         0.407         0.375   \n",
       "\n",
       "    ttest direction  \n",
       "0  0.1260         U  \n",
       "1  0.1460         U  \n",
       "2  0.0761         U  \n",
       "3  0.1640         U  \n",
       "4  0.0927         U  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# uncomment below line if you want to drop unwanted columns\n",
    "#df = df.drop(['startCpG', 'endCpG', 'len_bp', 'region', 'tg_mean', 'bg_mean', 'delta_quants', 'delta_maxmin'], axis=1)\n",
    "\n",
    "# remove CpGs from lenCpG values dataframe\n",
    "df['lenCpG'] = df['lenCpG'].apply(lambda x: x.split(\"CpGs\")[0])\n",
    "\n",
    "# remove bp from bp values in dataframe\n",
    "df['bp'] = df['bp'].apply(lambda x: x.split(\"bp\")[0])\n",
    "\n",
    "df.rename(columns = {'#chr':'chr'}, inplace = True)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine if there is overlap between genes and DMRs\n",
    "# query and reference\n",
    "def overlap(q_st, q_end, ref_st, ref_end):\n",
    "    overlap  = min(q_end, ref_end)-max(q_st, ref_st)\n",
    "    return overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gencode_annotate_loci(a, tb):\n",
    "    genes = []\n",
    "\n",
    "    try:\n",
    "        for region in tb.fetch(a['chr'], int(a['start']), int(a['end'])):\n",
    "            if region:\n",
    "                r = region.split('\\t')\n",
    "                overlap_len = overlap(int(a['start']), int(a['end']), int(r[1]), int(r[2]))\n",
    "                \n",
    "                ### Percentage of the input region that overlap with the feature\n",
    "                ret_val = '{}({})'.format(r[4], np.round(overlap_len/float(int(a['end'])-int(a['start']))*100, 2)) \n",
    "                \n",
    "                ### Percentage of the feature that overlaps with the region\n",
    "                # If you want to find the fraction of the GENCODE gene that overlaps with the input interval \n",
    "                # (e.g. overlap_len/float(int(r[2])-int(r[1]))*100).\n",
    "                # overlap_percent = '{}({})'.format(r[4], np.round(overlap_len/float(int(r[2])-int(r[1]))*100, 2)\n",
    "\n",
    "                genes.append(ret_val)\n",
    "\n",
    "                feature = r[3] \n",
    "                gene_type = r[5]\n",
    "                gene_level = r[6]\n",
    "                exon_id = r[7]\n",
    "                exon_number = r[8]\n",
    "                transcript_name = r[9]\n",
    "                transcript_type = r[10]\n",
    "                 \n",
    "\n",
    "        if len(genes)>0:\n",
    "            return [\";\".join(genes), feature, gene_type, gene_level, exon_id, exon_number, transcript_name, transcript_type]\n",
    "        else:\n",
    "            return [\"NA(0)\", np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan]\n",
    "    except ValueError:\n",
    "        return [\"NA(0)\", np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan]\n",
    "    \n",
    "# # loci is a row in a dataframe with columns chr, start, end ['#chr', 'start', 'end']\n",
    "# # tabix is the indexed annotation file as a pysam.TabixFile object\n",
    "# def gencode_annotate_loci(loci, tabix):\n",
    "#     genes = []\n",
    "#     try:\n",
    "#         # reads the tabix file\n",
    "#         for region in tabix.fetch(loci['chr'], int(loci['start']), int(loci['end'])):\n",
    "#             if region:\n",
    "#                 r = region.split('\\t')\n",
    "\n",
    "#                 ### Length of the overlap between the query interval and the gencode gene\n",
    "#                 # \n",
    "#                 overlap_len = overlap(int(loci['start']), int(loci['end']), int(r[1]), int(r[2]))\n",
    "\n",
    "#                 ### Percentage of the query interval that overlap with the gencode gene\n",
    "#                 # If you want to find the fraction of the GENCODE gene that overlaps with the input interval \n",
    "#                 # (e.g. overlap_len/float(int(r[2])-int(r[1]))*100).\n",
    "#                 query_overlap_percent = '{}({})'.format(r[3], np.round(overlap_len/float(int(loci['end'])-int(loci['start']))*100, 2)) \n",
    "#                 genes.append(query_overlap_percent)\n",
    "#         if len(genes)>0:\n",
    "#             return \";\".join(genes)\n",
    "#         else:\n",
    "#             return \"NA(0)\"\n",
    "#     except ValueError:\n",
    "#         return \"NA(0)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate dmrs using gencode_annotate_loci function and gencode_v44 annotation file\n",
    "# returns genes, feature, gene_type, gene_level, exon_id, exon_number, transcript_name, transcript_type that overlap with DMRs\n",
    "# returns amount of overlap as a percentage under gene column as gene-name(overlap %)\n",
    "df[['gene', 'feature', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type']] = df.apply(lambda x: gencode_annotate_loci(x[['chr', 'start', 'end']], gencode_v44), axis=1, result_type='expand')\n",
    "\n",
    "# replace empty strings with NaN\n",
    "df = df.replace(r'^\\s*$', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0);ENSG00000238009(100.0)</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0);ENSG00000238009(100.0)</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0);ENSG00000238009(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CICP27-201</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000228463(100.0);ENSG00000228463(100.0)</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000441866</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "      <td>NA(0)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2630</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000230021(100.0);ENSG00000230021(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000423796</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>598963</td>\n",
       "      <td>599165</td>\n",
       "      <td>5342</td>\n",
       "      <td>5347</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:598963-599165</td>\n",
       "      <td>5</td>\n",
       "      <td>202</td>\n",
       "      <td>0.293</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1310</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000230021(100.0);ENSG00000230021(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000634337</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>603632</td>\n",
       "      <td>604015</td>\n",
       "      <td>5406</td>\n",
       "      <td>5412</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:603632-604015</td>\n",
       "      <td>6</td>\n",
       "      <td>383</td>\n",
       "      <td>0.300</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0976</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000230021(100.0);ENSG00000230021(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000634337</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>605175</td>\n",
       "      <td>605246</td>\n",
       "      <td>5438</td>\n",
       "      <td>5446</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:605175-605246</td>\n",
       "      <td>8</td>\n",
       "      <td>71</td>\n",
       "      <td>0.454</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0957</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000230021(100.0);ENSG00000230021(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000634337</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>605698</td>\n",
       "      <td>607036</td>\n",
       "      <td>5475</td>\n",
       "      <td>5494</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:605698-607036</td>\n",
       "      <td>19</td>\n",
       "      <td>1338</td>\n",
       "      <td>0.290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1700</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000230021(100.0);ENSG00000230021(100.0);...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000634337</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "5  chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "6  chr1  598963  599165      5342    5347  colon_cancer  chr1:598963-599165   \n",
       "7  chr1  603632  604015      5406    5412  colon_cancer  chr1:603632-604015   \n",
       "8  chr1  605175  605246      5438    5446  colon_cancer  chr1:605175-605246   \n",
       "9  chr1  605698  607036      5475    5494  colon_cancer  chr1:605698-607036   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...   ttest  direction  \\\n",
       "0      6   416    0.311  ...  0.1260          U   \n",
       "1      7   468    0.260  ...  0.1460          U   \n",
       "2      8   206    0.633  ...  0.0761          U   \n",
       "3     11  1811    0.236  ...  0.1640          U   \n",
       "4     22  1169    0.435  ...  0.0927          U   \n",
       "5      7   193    0.122  ...  0.2630          U   \n",
       "6      5   202    0.293  ...  0.1310          U   \n",
       "7      6   383    0.300  ...  0.0976          U   \n",
       "8      8    71    0.454  ...  0.0957          U   \n",
       "9     19  1338    0.290  ...  0.1700          U   \n",
       "\n",
       "                                                gene     feature  \\\n",
       "0      ENSG00000238009(100.0);ENSG00000238009(100.0)        gene   \n",
       "1      ENSG00000238009(100.0);ENSG00000238009(100.0)        gene   \n",
       "2  ENSG00000238009(100.0);ENSG00000238009(100.0);...  transcript   \n",
       "3      ENSG00000228463(100.0);ENSG00000228463(100.0)  transcript   \n",
       "4                                              NA(0)         NaN   \n",
       "5  ENSG00000230021(100.0);ENSG00000230021(100.0);...  transcript   \n",
       "6  ENSG00000230021(100.0);ENSG00000230021(100.0);...  transcript   \n",
       "7  ENSG00000230021(100.0);ENSG00000230021(100.0);...  transcript   \n",
       "8  ENSG00000230021(100.0);ENSG00000230021(100.0);...  transcript   \n",
       "9  ENSG00000230021(100.0);ENSG00000230021(100.0);...  transcript   \n",
       "\n",
       "                          gene_type gene_level exon_id exon_number  \\\n",
       "0                            lncRNA          2     NaN         NaN   \n",
       "1                            lncRNA          2     NaN         NaN   \n",
       "2              processed_pseudogene          1     NaN         NaN   \n",
       "3  transcribed_processed_pseudogene          2     NaN         NaN   \n",
       "4                               NaN        NaN     NaN         NaN   \n",
       "5                            lncRNA          2     NaN         NaN   \n",
       "6  transcribed_processed_pseudogene          2     NaN         NaN   \n",
       "7  transcribed_processed_pseudogene          2     NaN         NaN   \n",
       "8  transcribed_processed_pseudogene          2     NaN         NaN   \n",
       "9  transcribed_processed_pseudogene          2     NaN         NaN   \n",
       "\n",
       "   transcript_name       transcript_type  \n",
       "0              NaN                   NaN  \n",
       "1              NaN                   NaN  \n",
       "2       CICP27-201  processed_pseudogene  \n",
       "3  ENST00000441866  processed_transcript  \n",
       "4              NaN                   NaN  \n",
       "5  ENST00000423796                lncRNA  \n",
       "6  ENST00000634337  processed_transcript  \n",
       "7  ENST00000634337  processed_transcript  \n",
       "8  ENST00000634337  processed_transcript  \n",
       "9  ENST00000634337  processed_transcript  \n",
       "\n",
       "[10 rows x 24 columns]"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "## skip if you want to keep regions with no genes\n",
    "## Transforms the data frame and removes anything that returned NA\n",
    "## Remove all the intervals that do not overlap with genes\n",
    "#df = df[df['genes'] != \"NA(0)\"].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region', 'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants', 'delta_maxmin', 'ttest', 'direction', 'gene', 'feature', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type']\n"
     ]
    }
   ],
   "source": [
    "column_names = list(df.columns.values)\n",
    "print(column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_rows = []\n",
    "for i,r in df.iterrows():\n",
    "    g_list = r['gene'].split(\";\")\n",
    "    for g in g_list:\n",
    "        g = g.replace(\" \",\"\")\n",
    "        new_rows.append(np.append(r[['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region', 'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants', 'delta_maxmin', 'ttest', 'direction', 'gene', 'feature', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type']].values, g))\n",
    "\n",
    "print(new_rows[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[array(['chr1', 92977, 93393, 1045, 1051, 'colon_cancer',\n",
      "       'chr1:92977-93393', '6', '416', 0.311, 0.724, 0.413, 0.372, 0.33,\n",
      "       0.126, 'U', 'ENSG00000238009(100.0);ENSG00000238009(100.0)',\n",
      "       'gene', 'lncRNA', '2', nan, nan, nan, nan,\n",
      "       'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 92977, 93393, 1045, 1051, 'colon_cancer',\n",
      "       'chr1:92977-93393', '6', '416', 0.311, 0.724, 0.413, 0.372, 0.33,\n",
      "       0.126, 'U', 'ENSG00000238009(100.0);ENSG00000238009(100.0)',\n",
      "       'gene', 'lncRNA', '2', nan, nan, nan, nan,\n",
      "       'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 99252, 99720, 1094, 1101, 'colon_cancer',\n",
      "       'chr1:99252-99720', '7', '468', 0.26, 0.667, 0.407, 0.36, 0.314,\n",
      "       0.146, 'U', 'ENSG00000238009(100.0);ENSG00000238009(100.0)',\n",
      "       'gene', 'lncRNA', '2', nan, nan, nan, nan,\n",
      "       'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 99252, 99720, 1094, 1101, 'colon_cancer',\n",
      "       'chr1:99252-99720', '7', '468', 0.26, 0.667, 0.407, 0.36, 0.314,\n",
      "       0.146, 'U', 'ENSG00000238009(100.0);ENSG00000238009(100.0)',\n",
      "       'gene', 'lncRNA', '2', nan, nan, nan, nan,\n",
      "       'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U',\n",
      "       'ENSG00000238009(100.0);ENSG00000238009(100.0);CICP27(100.0);CICP27(100.0);CICP27(100.0)',\n",
      "       'transcript', 'processed_pseudogene', '1', nan, nan, 'CICP27-201',\n",
      "       'processed_pseudogene', 'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U',\n",
      "       'ENSG00000238009(100.0);ENSG00000238009(100.0);CICP27(100.0);CICP27(100.0);CICP27(100.0)',\n",
      "       'transcript', 'processed_pseudogene', '1', nan, nan, 'CICP27-201',\n",
      "       'processed_pseudogene', 'ENSG00000238009(100.0)'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U',\n",
      "       'ENSG00000238009(100.0);ENSG00000238009(100.0);CICP27(100.0);CICP27(100.0);CICP27(100.0)',\n",
      "       'transcript', 'processed_pseudogene', '1', nan, nan, 'CICP27-201',\n",
      "       'processed_pseudogene', 'CICP27(100.0)'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U',\n",
      "       'ENSG00000238009(100.0);ENSG00000238009(100.0);CICP27(100.0);CICP27(100.0);CICP27(100.0)',\n",
      "       'transcript', 'processed_pseudogene', '1', nan, nan, 'CICP27-201',\n",
      "       'processed_pseudogene', 'CICP27(100.0)'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U',\n",
      "       'ENSG00000238009(100.0);ENSG00000238009(100.0);CICP27(100.0);CICP27(100.0);CICP27(100.0)',\n",
      "       'transcript', 'processed_pseudogene', '1', nan, nan, 'CICP27-201',\n",
      "       'processed_pseudogene', 'CICP27(100.0)'], dtype=object), array(['chr1', 275815, 277626, 2790, 2801, 'colon_cancer',\n",
      "       'chr1:275815-277626', '11', '1811', 0.236, 0.625, 0.389, 0.382,\n",
      "       0.375, 0.164, 'U', 'ENSG00000228463(100.0);ENSG00000228463(100.0)',\n",
      "       'transcript', 'transcribed_processed_pseudogene', '2', nan, nan,\n",
      "       'ENST00000441866', 'processed_transcript',\n",
      "       'ENSG00000228463(100.0)'], dtype=object)]\n"
     ]
    }
   ],
   "source": [
    "print(new_rows[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "160274"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(new_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>transcript</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CICP27-201</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "2  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "3  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "4  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "\n",
       "  lenCpG   bp  tg_mean  ...  direction     feature             gene_type  \\\n",
       "0      6  416    0.311  ...          U        gene                lncRNA   \n",
       "1      6  416    0.311  ...          U        gene                lncRNA   \n",
       "2      7  468    0.260  ...          U        gene                lncRNA   \n",
       "3      7  468    0.260  ...          U        gene                lncRNA   \n",
       "4      8  206    0.633  ...          U  transcript  processed_pseudogene   \n",
       "\n",
       "   gene_level  exon_id exon_number transcript_name       transcript_type  \\\n",
       "0           2      NaN         NaN             NaN                   NaN   \n",
       "1           2      NaN         NaN             NaN                   NaN   \n",
       "2           2      NaN         NaN             NaN                   NaN   \n",
       "3           2      NaN         NaN             NaN                   NaN   \n",
       "4           1      NaN         NaN      CICP27-201  processed_pseudogene   \n",
       "\n",
       "         gene_name gene_coverage  \n",
       "0  ENSG00000238009         100.0  \n",
       "1  ENSG00000238009         100.0  \n",
       "2  ENSG00000238009         100.0  \n",
       "3  ENSG00000238009         100.0  \n",
       "4  ENSG00000238009         100.0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_perGene = pd.DataFrame(new_rows, columns=['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region', 'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants', 'delta_maxmin', 'ttest', 'direction', 'gene', 'feature', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type', 'gene_ID'])\n",
    "\n",
    "df_perGene['gene_name'] = df_perGene['gene_ID'].apply(lambda x: x.split(\"(\")[0])\n",
    "\n",
    "df_perGene['gene_coverage'] = df_perGene['gene_ID'].apply(lambda x: x.split(\"(\")[1].replace(\")\", \"\"))\n",
    "\n",
    "## drop the genes column\n",
    "df_perGene = df_perGene.drop([\"gene\", \"gene_ID\"], axis=1)\n",
    "\n",
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>feature</th>\n",
       "      <th>percent_of_region_overlapping_feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>gene</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CICP27-201</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "2  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "3  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "4  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "\n",
       "  lenCpG   bp  tg_mean  ...  direction     feature  \\\n",
       "0      6  416    0.311  ...          U        gene   \n",
       "1      6  416    0.311  ...          U        gene   \n",
       "2      7  468    0.260  ...          U        gene   \n",
       "3      7  468    0.260  ...          U        gene   \n",
       "4      8  206    0.633  ...          U  transcript   \n",
       "\n",
       "   percent_of_region_overlapping_feature        gene_name  \\\n",
       "0                                  100.0  ENSG00000238009   \n",
       "1                                  100.0  ENSG00000238009   \n",
       "2                                  100.0  ENSG00000238009   \n",
       "3                                  100.0  ENSG00000238009   \n",
       "4                                  100.0  ENSG00000238009   \n",
       "\n",
       "              gene_type gene_level exon_id exon_number transcript_name  \\\n",
       "0                lncRNA          2     NaN         NaN             NaN   \n",
       "1                lncRNA          2     NaN         NaN             NaN   \n",
       "2                lncRNA          2     NaN         NaN             NaN   \n",
       "3                lncRNA          2     NaN         NaN             NaN   \n",
       "4  processed_pseudogene          1     NaN         NaN      CICP27-201   \n",
       "\n",
       "        transcript_type  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4  processed_pseudogene  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_perGene.rename(columns = {'gene_coverage':'percent_region_overlapping_feature'}, inplace = True)\n",
    "# note to self add percent feature overlapping region\n",
    "\n",
    "df_perGene = df_perGene.loc[:,['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region', 'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants', 'delta_maxmin', 'ttest', 'direction', 'feature', 'percent_of_region_overlapping_feature', 'gene_name', 'gene_type', 'gene_level', 'exon_id', 'exon_number', 'transcript_name', 'transcript_type']]\n",
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>feature</th>\n",
       "      <th>percent_of_region_overlapping_feature</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>160004</th>\n",
       "      <td>chrX</td>\n",
       "      <td>75064748</td>\n",
       "      <td>75066000</td>\n",
       "      <td>28572897</td>\n",
       "      <td>28572903</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:75064748-75066000</td>\n",
       "      <td>6</td>\n",
       "      <td>1252</td>\n",
       "      <td>0.706</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>13.66</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001713613.1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>ABCB7-213</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160005</th>\n",
       "      <td>chrX</td>\n",
       "      <td>75064748</td>\n",
       "      <td>75066000</td>\n",
       "      <td>28572897</td>\n",
       "      <td>28572903</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:75064748-75066000</td>\n",
       "      <td>6</td>\n",
       "      <td>1252</td>\n",
       "      <td>0.706</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>13.66</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001713613.1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>ABCB7-213</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160006</th>\n",
       "      <td>chrX</td>\n",
       "      <td>75064748</td>\n",
       "      <td>75066000</td>\n",
       "      <td>28572897</td>\n",
       "      <td>28572903</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:75064748-75066000</td>\n",
       "      <td>6</td>\n",
       "      <td>1252</td>\n",
       "      <td>0.706</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>13.66</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001713613.1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>ABCB7-213</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160012</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160013</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160014</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160015</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>12.15</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160016</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>12.15</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160017</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>12.15</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160018</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85345669</td>\n",
       "      <td>85347167</td>\n",
       "      <td>28633548</td>\n",
       "      <td>28633554</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:85345669-85347167</td>\n",
       "      <td>6</td>\n",
       "      <td>1498</td>\n",
       "      <td>0.771</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>12.15</td>\n",
       "      <td>POF1B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000847202.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>POF1B-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160107</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160108</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160109</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160110</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160111</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160112</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160113</th>\n",
       "      <td>chrX</td>\n",
       "      <td>115233999</td>\n",
       "      <td>115234037</td>\n",
       "      <td>28826498</td>\n",
       "      <td>28826503</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:115233999-115234037</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.627</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>LRCH2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003847550.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LRCH2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160146</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134172559</td>\n",
       "      <td>134172612</td>\n",
       "      <td>28972663</td>\n",
       "      <td>28972668</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:134172559-134172612</td>\n",
       "      <td>5</td>\n",
       "      <td>53</td>\n",
       "      <td>0.726</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000283638</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004013086.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENST00000710741</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160147</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134172559</td>\n",
       "      <td>134172612</td>\n",
       "      <td>28972663</td>\n",
       "      <td>28972668</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:134172559-134172612</td>\n",
       "      <td>5</td>\n",
       "      <td>53</td>\n",
       "      <td>0.726</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000283638</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004013086.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENST00000710741</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160148</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134172559</td>\n",
       "      <td>134172612</td>\n",
       "      <td>28972663</td>\n",
       "      <td>28972668</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:134172559-134172612</td>\n",
       "      <td>5</td>\n",
       "      <td>53</td>\n",
       "      <td>0.726</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000283638</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004013086.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENST00000710741</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160239</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156016784</td>\n",
       "      <td>156016903</td>\n",
       "      <td>29175063</td>\n",
       "      <td>29175074</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:156016784-156016903</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>0.682</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000270726</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160240</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156016784</td>\n",
       "      <td>156016903</td>\n",
       "      <td>29175063</td>\n",
       "      <td>29175074</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:156016784-156016903</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>0.682</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>100.0</td>\n",
       "      <td>ENSG00000270726</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160241</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156016784</td>\n",
       "      <td>156016903</td>\n",
       "      <td>29175063</td>\n",
       "      <td>29175074</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:156016784-156016903</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>0.682</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>44.54</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160242</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156016784</td>\n",
       "      <td>156016903</td>\n",
       "      <td>29175063</td>\n",
       "      <td>29175074</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:156016784-156016903</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>0.682</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>44.54</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160243</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156016784</td>\n",
       "      <td>156016903</td>\n",
       "      <td>29175063</td>\n",
       "      <td>29175074</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chrX:156016784-156016903</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>0.682</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>exon</td>\n",
       "      <td>44.54</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001772131.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WASIR1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end  startCpG    endCpG        target  \\\n",
       "160004  chrX   75064748   75066000  28572897  28572903  colon_cancer   \n",
       "160005  chrX   75064748   75066000  28572897  28572903  colon_cancer   \n",
       "160006  chrX   75064748   75066000  28572897  28572903  colon_cancer   \n",
       "160012  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160013  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160014  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160015  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160016  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160017  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160018  chrX   85345669   85347167  28633548  28633554  colon_cancer   \n",
       "160107  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160108  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160109  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160110  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160111  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160112  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160113  chrX  115233999  115234037  28826498  28826503  colon_cancer   \n",
       "160146  chrX  134172559  134172612  28972663  28972668  colon_cancer   \n",
       "160147  chrX  134172559  134172612  28972663  28972668  colon_cancer   \n",
       "160148  chrX  134172559  134172612  28972663  28972668  colon_cancer   \n",
       "160239  chrX  156016784  156016903  29175063  29175074  colon_cancer   \n",
       "160240  chrX  156016784  156016903  29175063  29175074  colon_cancer   \n",
       "160241  chrX  156016784  156016903  29175063  29175074  colon_cancer   \n",
       "160242  chrX  156016784  156016903  29175063  29175074  colon_cancer   \n",
       "160243  chrX  156016784  156016903  29175063  29175074  colon_cancer   \n",
       "\n",
       "                          region lenCpG    bp  tg_mean  ...  direction  \\\n",
       "160004    chrX:75064748-75066000      6  1252    0.706  ...          M   \n",
       "160005    chrX:75064748-75066000      6  1252    0.706  ...          M   \n",
       "160006    chrX:75064748-75066000      6  1252    0.706  ...          M   \n",
       "160012    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160013    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160014    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160015    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160016    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160017    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160018    chrX:85345669-85347167      6  1498    0.771  ...          M   \n",
       "160107  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160108  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160109  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160110  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160111  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160112  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160113  chrX:115233999-115234037      5    38    0.627  ...          M   \n",
       "160146  chrX:134172559-134172612      5    53    0.726  ...          M   \n",
       "160147  chrX:134172559-134172612      5    53    0.726  ...          M   \n",
       "160148  chrX:134172559-134172612      5    53    0.726  ...          M   \n",
       "160239  chrX:156016784-156016903     11   119    0.682  ...          M   \n",
       "160240  chrX:156016784-156016903     11   119    0.682  ...          M   \n",
       "160241  chrX:156016784-156016903     11   119    0.682  ...          M   \n",
       "160242  chrX:156016784-156016903     11   119    0.682  ...          M   \n",
       "160243  chrX:156016784-156016903     11   119    0.682  ...          M   \n",
       "\n",
       "        feature  percent_of_region_overlapping_feature        gene_name  \\\n",
       "160004     exon                                  13.66            ABCB7   \n",
       "160005     exon                                  13.66            ABCB7   \n",
       "160006     exon                                  13.66            ABCB7   \n",
       "160012     exon                                  100.0            POF1B   \n",
       "160013     exon                                  100.0            POF1B   \n",
       "160014     exon                                  100.0            POF1B   \n",
       "160015     exon                                  12.15            POF1B   \n",
       "160016     exon                                  12.15            POF1B   \n",
       "160017     exon                                  12.15            POF1B   \n",
       "160018     exon                                  12.15            POF1B   \n",
       "160107     exon                                  100.0            LRCH2   \n",
       "160108     exon                                  100.0            LRCH2   \n",
       "160109     exon                                  100.0            LRCH2   \n",
       "160110     exon                                  100.0            LRCH2   \n",
       "160111     exon                                  100.0            LRCH2   \n",
       "160112     exon                                  100.0            LRCH2   \n",
       "160113     exon                                  100.0            LRCH2   \n",
       "160146     exon                                  100.0  ENSG00000283638   \n",
       "160147     exon                                  100.0  ENSG00000283638   \n",
       "160148     exon                                  100.0  ENSG00000283638   \n",
       "160239     exon                                  100.0  ENSG00000270726   \n",
       "160240     exon                                  100.0  ENSG00000270726   \n",
       "160241     exon                                  44.54           WASIR1   \n",
       "160242     exon                                  44.54           WASIR1   \n",
       "160243     exon                                  44.54           WASIR1   \n",
       "\n",
       "             gene_type gene_level            exon_id exon_number  \\\n",
       "160004  protein_coding          2  ENSE00001713613.1        14.0   \n",
       "160005  protein_coding          2  ENSE00001713613.1        14.0   \n",
       "160006  protein_coding          2  ENSE00001713613.1        14.0   \n",
       "160012  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160013  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160014  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160015  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160016  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160017  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160018  protein_coding          2  ENSE00000847202.1         6.0   \n",
       "160107  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160108  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160109  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160110  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160111  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160112  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160113  protein_coding          2  ENSE00003847550.1         1.0   \n",
       "160146          lncRNA          2  ENSE00004013086.1         2.0   \n",
       "160147          lncRNA          2  ENSE00004013086.1         2.0   \n",
       "160148          lncRNA          2  ENSE00004013086.1         2.0   \n",
       "160239          lncRNA          2  ENSE00001772131.3         1.0   \n",
       "160240          lncRNA          2  ENSE00001772131.3         1.0   \n",
       "160241          lncRNA          2  ENSE00001772131.3         1.0   \n",
       "160242          lncRNA          2  ENSE00001772131.3         1.0   \n",
       "160243          lncRNA          2  ENSE00001772131.3         1.0   \n",
       "\n",
       "        transcript_name transcript_type  \n",
       "160004        ABCB7-213  protein_coding  \n",
       "160005        ABCB7-213  protein_coding  \n",
       "160006        ABCB7-213  protein_coding  \n",
       "160012        POF1B-202  protein_coding  \n",
       "160013        POF1B-202  protein_coding  \n",
       "160014        POF1B-202  protein_coding  \n",
       "160015        POF1B-202  protein_coding  \n",
       "160016        POF1B-202  protein_coding  \n",
       "160017        POF1B-202  protein_coding  \n",
       "160018        POF1B-202  protein_coding  \n",
       "160107        LRCH2-201  protein_coding  \n",
       "160108        LRCH2-201  protein_coding  \n",
       "160109        LRCH2-201  protein_coding  \n",
       "160110        LRCH2-201  protein_coding  \n",
       "160111        LRCH2-201  protein_coding  \n",
       "160112        LRCH2-201  protein_coding  \n",
       "160113        LRCH2-201  protein_coding  \n",
       "160146  ENST00000710741          lncRNA  \n",
       "160147  ENST00000710741          lncRNA  \n",
       "160148  ENST00000710741          lncRNA  \n",
       "160239       WASIR1-201          lncRNA  \n",
       "160240       WASIR1-201          lncRNA  \n",
       "160241       WASIR1-201          lncRNA  \n",
       "160242       WASIR1-201          lncRNA  \n",
       "160243       WASIR1-201          lncRNA  \n",
       "\n",
       "[25 rows x 25 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# return rows where feature is exon in df_perGene\n",
    "df_perGene[df_perGene.feature=='exon'].tail(25)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv has been created\n"
     ]
    }
   ],
   "source": [
    "df_perGene.to_csv(output_file, index=False, header=True, sep=\"\\t\")\n",
    "\n",
    "print( output_file   + \" has been created\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "annot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
