{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Info:\n",
    "    # Script name: 0_create_annotation_references_simplified.ipynb\n",
    "    # Purpose: Create annotation references for DMR analysis\n",
    "    # Author: Meghan M. Sleeper\n",
    "    # Input: gencode.v44.chr_patch_hapl_scaff.basic.annotation.gff3\n",
    "    # Outputs: \n",
    "        # gencode.v44.annotation.genes.txt\n",
    "        # gencode.v44.annotation.genes.txt.sorted.formatted.bed \n",
    "        # gencode.v44.annotation.genes.txt.sorted.formatted.bed.gz\n",
    "        # gencode.v44.annotation.all.txt\n",
    "        # gencode.v44.annotation.all.txt.sorted.formatted.bed \n",
    "        # gencode.v44.annotation.all.txt.sorted.formatted.bed.gz\n",
    "\n",
    "# Notes:\n",
    "    # Modified from: https://medium.com/intothegenomics/annotate-genes-and-genomic-coordinates-using-python-9259efa6ffc2\n",
    "\n",
    "# Reminders to self:\n",
    "    # gencode_exons_codons_utrs_cds.head(10)\n",
    "    # gencode.feature.unique()\n",
    "    # gencode.head()\n",
    "    # gencode[gencode.feature == 'gene'].head(10)\n",
    "    # list(gencode.attribute)[0:10]\n",
    "    # gencode[gencode.attribute.str.contains('gene_type')].head(10)\n",
    "    # magic commands: \n",
    "        # %reset can remove variables from memory\n",
    "\n",
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "\n",
    "# Assign file paths\n",
    "# unzipped gttf annotation reference\n",
    "gencode_annotation_file = \"/Users/meghansleeper/Desktop/farm-files/dmr-workflow/supplementary/gencode.v44.chr_patch_hapl_scaff.basic.annotation.gff3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gunzip: /Users/meghansleeper/Desktop/farm-files/dmr-workflow/supplementary/gencode.v44.chr_patch_hapl_scaff.basic.annotation.gff3: unknown suffix -- ignored\n"
     ]
    }
   ],
   "source": [
    "# %%bash -s gencode_annotation_file \n",
    "\n",
    "# # add code to unzip gencode_annotation_file if it is a .gz file\n",
    "# if gencode_annotation_file.endswith(\".gz\"):\n",
    "#     with gzip.open(gencode_annotation_file, 'rb') as f_in:\n",
    "#         with open(gencode_annotation_file[:-3], 'wb') as f_out:\n",
    "#             shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dataframe from gencode gtf annotation file\n",
    "gencode = pd.read_table(gencode_annotation_file, \n",
    "                        comment=\"#\", \n",
    "                        sep = \"\\t\", \n",
    "                        names = ['seqname', 'source', 'feature', \n",
    "                                'start' , 'end', 'score', 'strand', \n",
    "                                'frame', 'attribute'\n",
    "                                ]\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating dataframes for grouped features that share the same attribute information to make parsing attributes easier\n",
    "  # see https://www.gencodegenes.org/pages/data_format.html for attributes associated with each feature\n",
    "    # gencode.feature.unique() returns:\n",
    "        #    'gene', 'transcript', 'exon', 'CDS', \n",
    "        #    'start_codon', 'stop_codon',\n",
    "        #    'five_prime_UTR', 'three_prime_UTR',\n",
    "        #    'stop_codon_redefined_as_selenocysteine'\n",
    "\n",
    "gencode_genes = gencode[(gencode.feature == \"gene\")][\n",
    "    ['seqname', 'start', 'end', 'attribute','feature', \n",
    "     'strand', 'source']].copy().reset_index().drop('index', axis=1)\n",
    "\n",
    "gencode_exons_codons_utrs_cds = gencode[\n",
    "    (gencode.feature == \"exon\") | \n",
    "    (gencode.feature == \"start_codon\") | \n",
    "    (gencode.feature == \"stop_codon\") | \n",
    "    (gencode.feature == \"five_prime_UTR\") | \n",
    "    (gencode.feature == \"three_prime_UTR\") | \n",
    "    (gencode.feature == \"CDS\")\n",
    "    ][['seqname', 'start', 'end', 'attribute', 'feature', 'strand', 'source']\n",
    "      ].copy().reset_index().drop('index', axis=1)\n",
    "\n",
    "gencode_transcripts_selenocyteines = gencode[\n",
    "    (gencode.feature == \"transcript\") | \n",
    "    (gencode.feature == \"stop_codon_redefined_as_selenocysteine\")\n",
    "    ][['seqname', 'start', 'end', 'attribute', 'feature', 'strand', 'source']\n",
    "      ].copy().reset_index().drop('index', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=ENST00000466557.6;Parent=ENSG00000241860.7;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:1;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=1;exon_id=ENSE00001947154.2;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:2;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=2;exon_id=ENSE00001890990.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:3;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=3;exon_id=ENSE00001922566.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:4;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=4;exon_id=ENSE00001947087.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:5;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=5;exon_id=ENSE00001949945.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:6;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=6;exon_id=ENSE00001862111.2;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:7;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=7;exon_id=ENSE00001799309.2;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=exon:ENST00000466557.6:8;Parent=ENST00000466557.6;gene_id=ENSG00000241860.7;transcript_id=ENST00000466557.6;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000466557;exon_number=8;exon_id=ENSE00001853409.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000007037.2',\n",
       " 'ID=ENST00000662089.1;Parent=ENSG00000241860.7;gene_id=ENSG00000241860.7;transcript_id=ENST00000662089.1;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000662089;level=2;tag=basic,TAGENE;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000528578.1',\n",
       " 'ID=exon:ENST00000662089.1:1;Parent=ENST00000662089.1;gene_id=ENSG00000241860.7;transcript_id=ENST00000662089.1;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000662089;exon_number=1;exon_id=ENSE00003878614.1;level=2;tag=basic,TAGENE;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000528578.1',\n",
       " 'ID=exon:ENST00000662089.1:2;Parent=ENST00000662089.1;gene_id=ENSG00000241860.7;transcript_id=ENST00000662089.1;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000662089;exon_number=2;exon_id=ENSE00003868429.1;level=2;tag=basic,TAGENE;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000528578.1',\n",
       " 'ID=exon:ENST00000662089.1:3;Parent=ENST00000662089.1;gene_id=ENSG00000241860.7;transcript_id=ENST00000662089.1;gene_type=lncRNA;gene_name=ENSG00000241860;transcript_type=lncRNA;transcript_name=ENST00000662089;exon_number=3;exon_id=ENSE00003873220.1;level=2;tag=basic,TAGENE;havana_gene=OTTHUMG00000002480.4;havana_transcript=OTTHUMT00000528578.1',\n",
       " 'ID=ENSG00000222623.1;gene_id=ENSG00000222623.1;gene_type=snRNA;gene_name=RNU6-1100P;level=3;hgnc_id=HGNC:48063',\n",
       " 'ID=ENST00000410691.1;Parent=ENSG00000222623.1;gene_id=ENSG00000222623.1;transcript_id=ENST00000410691.1;gene_type=snRNA;gene_name=RNU6-1100P;transcript_type=snRNA;transcript_name=RNU6-1100P-201;level=3;transcript_support_level=NA;hgnc_id=HGNC:48063;tag=basic,Ensembl_canonical',\n",
       " 'ID=exon:ENST00000410691.1:1;Parent=ENST00000410691.1;gene_id=ENSG00000222623.1;transcript_id=ENST00000410691.1;gene_type=snRNA;gene_name=RNU6-1100P;transcript_type=snRNA;transcript_name=RNU6-1100P-201;exon_number=1;exon_id=ENSE00001807945.1;level=3;transcript_support_level=NA;hgnc_id=HGNC:48063;tag=basic,Ensembl_canonical',\n",
       " 'ID=ENSG00000241599.1;gene_id=ENSG00000241599.1;gene_type=lncRNA;gene_name=ENSG00000241599;level=2;havana_gene=OTTHUMG00000002525.1',\n",
       " 'ID=ENST00000496488.1;Parent=ENSG00000241599.1;gene_id=ENSG00000241599.1;transcript_id=ENST00000496488.1;gene_type=lncRNA;gene_name=ENSG00000241599;transcript_type=lncRNA;transcript_name=ENST00000496488;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,dotter_confirmed,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002525.1;havana_transcript=OTTHUMT00000007169.1',\n",
       " 'ID=exon:ENST00000496488.1:1;Parent=ENST00000496488.1;gene_id=ENSG00000241599.1;transcript_id=ENST00000496488.1;gene_type=lncRNA;gene_name=ENSG00000241599;transcript_type=lncRNA;transcript_name=ENST00000496488;exon_number=1;exon_id=ENSE00001882727.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,dotter_confirmed,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002525.1;havana_transcript=OTTHUMT00000007169.1',\n",
       " 'ID=exon:ENST00000496488.1:2;Parent=ENST00000496488.1;gene_id=ENSG00000241599.1;transcript_id=ENST00000496488.1;gene_type=lncRNA;gene_name=ENSG00000241599;transcript_type=lncRNA;transcript_name=ENST00000496488;exon_number=2;exon_id=ENSE00001895361.1;level=2;transcript_support_level=5;tag=not_best_in_genome_evidence,dotter_confirmed,basic,Ensembl_canonical;havana_gene=OTTHUMG00000002525.1;havana_transcript=OTTHUMT00000007169.1',\n",
       " 'ID=ENSG00000279928.2;gene_id=ENSG00000279928.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;level=2;hgnc_id=HGNC:55080;havana_gene=OTTHUMG00000191962.1',\n",
       " 'ID=ENST00000624431.2;Parent=ENSG00000279928.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=exon:ENST00000624431.2:1;Parent=ENST00000624431.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;exon_number=1;exon_id=ENSE00003759020.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=exon:ENST00000624431.2:2;Parent=ENST00000624431.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;exon_number=2;exon_id=ENSE00003759581.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=exon:ENST00000624431.2:3;Parent=ENST00000624431.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;exon_number=3;exon_id=ENSE00003804405.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=exon:ENST00000624431.2:4;Parent=ENST00000624431.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;exon_number=4;exon_id=ENSE00003807458.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=exon:ENST00000624431.2:5;Parent=ENST00000624431.2;gene_id=ENSG00000279928.2;transcript_id=ENST00000624431.2;gene_type=unprocessed_pseudogene;gene_name=DDX11L17;transcript_type=unprocessed_pseudogene;transcript_name=DDX11L17-201;exon_number=5;exon_id=ENSE00003760199.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:55080;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191962.1;havana_transcript=OTTHUMT00000491420.1',\n",
       " 'ID=ENSG00000279457.4;gene_id=ENSG00000279457.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;level=2;hgnc_id=HGNC:53981;havana_gene=OTTHUMG00000191963.1',\n",
       " 'ID=ENST00000623083.4;Parent=ENSG00000279457.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:1;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=1;exon_id=ENSE00003755778.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:2;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=2;exon_id=ENSE00003755850.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:3;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=3;exon_id=ENSE00003755821.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:4;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=4;exon_id=ENSE00003759270.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:5;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=5;exon_id=ENSE00003757255.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:6;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=6;exon_id=ENSE00003755701.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:7;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=7;exon_id=ENSE00003759679.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:8;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=8;exon_id=ENSE00003757727.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:9;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=9;exon_id=ENSE00003757943.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=exon:ENST00000623083.4:10;Parent=ENST00000623083.4;gene_id=ENSG00000279457.4;transcript_id=ENST00000623083.4;gene_type=unprocessed_pseudogene;gene_name=WASH9P;transcript_type=unprocessed_pseudogene;transcript_name=WASH9P-201;exon_number=10;exon_id=ENSE00003755076.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:53981;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000191963.1;havana_transcript=OTTHUMT00000491421.1',\n",
       " 'ID=ENSG00000273874.1;gene_id=ENSG00000273874.1;gene_type=miRNA;gene_name=MIR6859-2;level=3;hgnc_id=HGNC:49954',\n",
       " 'ID=ENST00000612080.1;Parent=ENSG00000273874.1;gene_id=ENSG00000273874.1;transcript_id=ENST00000612080.1;gene_type=miRNA;gene_name=MIR6859-2;transcript_type=miRNA;transcript_name=MIR6859-2-201;level=3;transcript_support_level=NA;hgnc_id=HGNC:49954;tag=basic,Ensembl_canonical',\n",
       " 'ID=exon:ENST00000612080.1:1;Parent=ENST00000612080.1;gene_id=ENSG00000273874.1;transcript_id=ENST00000612080.1;gene_type=miRNA;gene_name=MIR6859-2;transcript_type=miRNA;transcript_name=MIR6859-2-201;exon_number=1;exon_id=ENSE00003737837.1;level=3;transcript_support_level=NA;hgnc_id=HGNC:49954;tag=basic,Ensembl_canonical',\n",
       " 'ID=ENSG00000228463.10;gene_id=ENSG00000228463.10;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;level=1;tag=pseudo_consens;havana_gene=OTTHUMG00000002552.3',\n",
       " 'ID=ENST00000442116.1;Parent=ENSG00000228463.10;gene_id=ENSG00000228463.10;transcript_id=ENST00000442116.1;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000442116;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000007244.1',\n",
       " 'ID=exon:ENST00000442116.1:1;Parent=ENST00000442116.1;gene_id=ENSG00000228463.10;transcript_id=ENST00000442116.1;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000442116;exon_number=1;exon_id=ENSE00001149252.5;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000007244.1',\n",
       " 'ID=exon:ENST00000442116.1:2;Parent=ENST00000442116.1;gene_id=ENSG00000228463.10;transcript_id=ENST00000442116.1;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000442116;exon_number=2;exon_id=ENSE00001677641.1;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000007244.1',\n",
       " 'ID=ENST00000441866.2;Parent=ENSG00000228463.10;gene_id=ENSG00000228463.10;transcript_id=ENST00000441866.2;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000441866;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000346905.2',\n",
       " 'ID=exon:ENST00000441866.2:1;Parent=ENST00000441866.2;gene_id=ENSG00000228463.10;transcript_id=ENST00000441866.2;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000441866;exon_number=1;exon_id=ENSE00001670287.1;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000346905.2',\n",
       " 'ID=exon:ENST00000441866.2:2;Parent=ENST00000441866.2;gene_id=ENSG00000228463.10;transcript_id=ENST00000441866.2;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000441866;exon_number=2;exon_id=ENSE00001729229.1;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000346905.2',\n",
       " 'ID=exon:ENST00000441866.2:3;Parent=ENST00000441866.2;gene_id=ENSG00000228463.10;transcript_id=ENST00000441866.2;gene_type=transcribed_processed_pseudogene;gene_name=ENSG00000228463;transcript_type=processed_transcript;transcript_name=ENST00000441866;exon_number=3;exon_id=ENSE00003789911.2;level=2;transcript_support_level=5;tag=basic;havana_gene=OTTHUMG00000002552.3;havana_transcript=OTTHUMT00000346905.2']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(gencode.attribute)[100:150]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions for parsing gtf files based on format of attribute column\n",
    "\n",
    "def gene_info(x):\n",
    "# Extract gene names, gene id gene_type, level, and gene_id for each gene\n",
    "# Use this function when targetting info on genes\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_id = list(filter(lambda x: 'gene_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_level = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    return (g_name, g_id, g_type, g_level)\n",
    "\n",
    "def all_info(x):\n",
    "# Extract gene names, gene id, gene_type, level, exon_id, exon_number, transcript_name, and transcript_type for each exon\n",
    "# Use this function when targetting info on exons, CDS, start_codon, stop_codon, five_prime_UTR, and three_prime_UTR\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_id = list(filter(lambda x: 'gene_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_level = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    exon_id = list(filter(lambda x: 'exon_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    exon_number = int(list(filter(lambda x: 'exon_number' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    transcript_name = list(filter(lambda x: 'transcript_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    transcript_type = list(filter(lambda x: 'transcript_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    return (g_name, g_id, g_type, g_level, exon_id, exon_number, transcript_name, transcript_type)\n",
    "    \n",
    "def transcript_info(x):\n",
    "# Extract gene names, gene id, gene_type, level, transcript_id, and transcript_type for each transcript\n",
    "# Use this function when targetting info on transcripts and stop_codon_redefined_as_selenocysteine\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_id = list(filter(lambda x: 'gene_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_level = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    transcript_name = list(filter(lambda x: 'transcript_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    transcript_type = list(filter(lambda x: 'transcript_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    return (g_name, g_id, g_type, g_level, transcript_name, transcript_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Using gene_info function on genes\n",
    "# extract gene_name, gene_id, gene_type, and gene_level from gencode_genes.attribute\n",
    "gencode_genes[\"gene_name\"], gencode_genes[\"gene_id\"], gencode_genes[\"gene_type\"], gencode_genes[\"gene_level\"] = zip(*gencode_genes.attribute.apply(lambda x: gene_info(x)))\n",
    "\n",
    "## Using all_info function on exons, start codons, stop codons, utrs, and cds\n",
    "# extract gene_name, gene_type, gene_level, exon_id, exon_number, transcript_name, and transcript_type from gencode_exons_codons_utrs_cds.attribute\n",
    "gencode_exons_codons_utrs_cds[\"gene_name\"], gencode_exons_codons_utrs_cds[\"gene_id\"], gencode_exons_codons_utrs_cds[\"gene_type\"], gencode_exons_codons_utrs_cds[\"gene_level\"], gencode_exons_codons_utrs_cds[\"exon_id\"], gencode_exons_codons_utrs_cds[\"exon_number\"], gencode_exons_codons_utrs_cds[\"transcript_name\"], gencode_exons_codons_utrs_cds[\"transcript_type\"] = zip(*gencode_exons_codons_utrs_cds.attribute.apply(lambda x: all_info(x)))\n",
    "\n",
    "## Using transcript_info function on transcripts and stop_codon_redefined_as_selenocysteine\n",
    "# extract gene_name, gene_type, gene_level, exon_id, exon_number, transcript_name, and transcript_type from gencode_transcripts_selenocyteines.attribute using transcript_info function\n",
    "gencode_transcripts_selenocyteines[\"gene_name\"], gencode_transcripts_selenocyteines[\"gene_id\"], gencode_transcripts_selenocyteines[\"gene_type\"], gencode_transcripts_selenocyteines[\"gene_level\"], gencode_transcripts_selenocyteines[\"transcript_name\"], gencode_transcripts_selenocyteines[\"transcript_type\"] = zip(*gencode_transcripts_selenocyteines.attribute.apply(lambda x: transcript_info(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gencode_genes.gene_type.unique() returns:\n",
    "    # array(['lncRNA', 'transcribed_unprocessed_pseudogene',\n",
    "    #        'unprocessed_pseudogene', 'miRNA', 'protein_coding',\n",
    "    #        'processed_pseudogene', 'snRNA',\n",
    "    #        'transcribed_processed_pseudogene', 'misc_RNA', 'TEC',\n",
    "    #        'transcribed_unitary_pseudogene', 'snoRNA', 'scaRNA',\n",
    "    #        'rRNA_pseudogene', 'unitary_pseudogene', 'pseudogene', 'rRNA',\n",
    "    #        'IG_V_pseudogene', 'scRNA', 'IG_V_gene', 'IG_C_gene', 'IG_J_gene',\n",
    "    #        'sRNA', 'ribozyme', 'translated_processed_pseudogene', 'vault_RNA',\n",
    "    #        'TR_C_gene', 'TR_J_gene', 'TR_V_gene', 'TR_V_pseudogene',\n",
    "    #        'TR_D_gene', 'IG_C_pseudogene', 'TR_J_pseudogene',\n",
    "    #        'IG_J_pseudogene', 'IG_D_gene', 'IG_pseudogene', 'artifact',\n",
    "    #        'Mt_tRNA', 'Mt_rRNA'], dtype=object)\n",
    "\n",
    "# For filtering out the genes that are protein coding into a new dataframe, uncomment the following line\n",
    "# gencode_genes_protein_coding = gencode_genes[gencode_genes['gene_type'] == 'protein_coding'].reset_index().drop('index', axis=1)\n",
    "\n",
    "# concatenating dataframes for all features into one datafram with NaNs for missing values where features don't have the same attributes\n",
    "gencode_all = pd.concat(\n",
    "    [gencode_genes, \n",
    "     gencode_exons_codons_utrs_cds, \n",
    "     gencode_transcripts_selenocyteines], \n",
    "     ignore_index=True\n",
    "     ).reset_index().drop('index', axis=1)\n",
    "\n",
    "# To confirm that all features are included in the new dataframe, uncomment the following line\n",
    "# gencode_all.feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>attribute</th>\n",
       "      <th>feature</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end                                          attribute  \\\n",
       "0    chr1  11869  14409  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...   \n",
       "1    chr1  12010  13670  ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...   \n",
       "2    chr1  14404  29570  ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...   \n",
       "3    chr1  17369  17436  ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...   \n",
       "4    chr1  29554  31109  ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...   \n",
       "\n",
       "  feature strand   source    gene_name            gene_id  \\\n",
       "0    gene      +   HAVANA      DDX11L2  ENSG00000290825.1   \n",
       "1    gene      +   HAVANA      DDX11L1  ENSG00000223972.6   \n",
       "2    gene      -   HAVANA       WASH7P  ENSG00000227232.5   \n",
       "3    gene      -  ENSEMBL    MIR6859-1  ENSG00000278267.1   \n",
       "4    gene      +   HAVANA  MIR1302-2HG  ENSG00000243485.5   \n",
       "\n",
       "                            gene_type  gene_level  \n",
       "0                              lncRNA           2  \n",
       "1  transcribed_unprocessed_pseudogene           2  \n",
       "2              unprocessed_pseudogene           2  \n",
       "3                               miRNA           3  \n",
       "4                              lncRNA           2  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['seqname',\n",
       " 'start',\n",
       " 'end',\n",
       " 'attribute',\n",
       " 'feature',\n",
       " 'strand',\n",
       " 'source',\n",
       " 'gene_name',\n",
       " 'gene_id',\n",
       " 'gene_type',\n",
       " 'gene_level',\n",
       " 'exon_id',\n",
       " 'exon_number',\n",
       " 'transcript_name',\n",
       " 'transcript_type']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(gencode_all.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving, sorting, and indexing new annotation reference files\n",
    "\n",
    "The gene annotation files will be saved in bed format with tab delimition and no header row. The rows will be sorted by chromosome, start, and end (in that order).\n",
    "\n",
    "`gencode_all` contains all feature types and will be saved  as `gencode.v44.annotation.all.sorted.formatted.bed` with the columns:\n",
    "    ```['seqname', 'start', 'end', 'attribute',\n",
    "    'feature', 'strand', 'source', 'gene_name',\n",
    "    'gene_id', 'gene_type', 'gene_level', 'exon_id',\n",
    "    'exon_number', 'transcript_name', 'transcript_type']```\n",
    "\n",
    "`gencode_gene` contains genes only and will be saved  as `gencode.v44.annotation.genes.sorted.formatted.bed` with the columns:\n",
    "    ```['seqname', 'start', 'end', 'attribute',\n",
    "    'feature', 'strand', 'source', 'gene_name',\n",
    "    'gene_id', 'gene_type', 'gene_level']```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Save the gencode dataframes into files to be sorted and converted to bed files\n",
    "gencode_all.to_csv('gencode.v44.annotation.all.tsv', index=False, header = False, sep=\"\\t\")\n",
    "gencode_genes.to_csv('gencode.v44.annotation.genes.tsv', index=False, header = False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s gencode.v44.annotation.all.tsv\n",
    "# Sort by chromosome, start, and end\n",
    "cut -f 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15 $1 | sort -k1,1 -k2,2n -k3,3n > gencode.v44.annotation.all.sorted.formatted.bed\n",
    "\n",
    "# Compress the bed file\n",
    "bgzip -f gencode.v44.annotation.all.sorted.formatted.bed\n",
    "\n",
    "# Index the bed file\n",
    "tabix -f -p bed gencode.v44.annotation.all.sorted.formatted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s gencode.v44.annotation.genes.tsv\n",
    "# Sort by chromosome, start, and end\n",
    "cut -f 1,2,3,4,5,6,7,8,9,10,11 $1 | sort -k1,1 -k2,2n -k3,3n > gencode.v44.annotation.genes.sorted.formatted.bed\n",
    "\n",
    "# Compress the bed file\n",
    "bgzip -f gencode.v44.annotation.genes.sorted.formatted.bed\n",
    "\n",
    "# Index the bed file\n",
    "tabix -f -p bed gencode.v44.annotation.genes.sorted.formatted.bed.gz"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "annot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
