{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conda environment to use: annot\n",
    "\n",
    "# References for sources used to write this script:\n",
    "#   - Gene annotation methods: https://medium.com/intothegenomics/annotate-genes-and-genomic-coordinates-using-python-9259efa6ffc2\n",
    "#   - Gencode annotation file format: https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "# figure out how to interpret the chromosome labels\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=\n",
    "\n",
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "\n",
    "# Assign directories and files to use for annotation references\n",
    "#annotation_file = \"/Users/meghansleeper/Python/deconvolution_projects/dmr_workflow/supplementary/annotation_beds/gencode.v44.annotation.all.sorted.formatted.bed.gz\"\n",
    "annotation_file_genes = \"/Users/meghansleeper/Python/deconvolution_projects/dmr_workflow/supplementary/annotation_beds/gencode.v44.annotation.genes.sorted.formatted.bed.gz\"\n",
    "annotation_file_regulatory = \"/Users/meghansleeper/Python/deconvolution_projects/dmr_workflow/supplementary/annotation_beds/ensembl.GRch38.p14.regulatory.sorted.formatted.bed.gz\"\n",
    "\n",
    "# DMR regions to annotate (tab-delimited bed formatted file)\n",
    "DMR_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers.tsv\"\n",
    "\n",
    "# Output file names\n",
    "output_file_per_gene = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_per_gene.tsv\"\n",
    "output_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv\"\n",
    "output_file_per_reg = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_per_reg.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference for genes only: \n",
    "# Read reference annotation file with tabix\n",
    "gencode_v44_genes = pysam.TabixFile(annotation_file_genes)\n",
    "# read in the annotation file as a pandas dataframe\n",
    "annot_df_gene = pd.read_csv(annotation_file_genes, sep='\\t', header=None, names=['chr', 'start', 'end', \n",
    "                                                                                'attribute', 'feature', 'strand', \n",
    "                                                                                'source', 'gene_name', 'gene_id', \n",
    "                                                                                'gene_type', 'gene_level'])\n",
    "# Reference for regulatory features from ensembl regulatory build:\n",
    "# Read reference annotation file with tabix\n",
    "regulatory_build = pysam.TabixFile(annotation_file_regulatory)\n",
    "annot_df_regulatory = pd.read_csv(annotation_file_regulatory, \n",
    "                                  sep='\\t', \n",
    "                                  header=0, \n",
    "                                  low_memory=False, \n",
    "                                  names=['chr', 'start', 'end',\n",
    "                                         'feature_type', 'feature_type_description'])\n",
    "\n",
    "# # Reference for all features: \n",
    "# # Read reference annotation file with tabix\n",
    "# gencode_v44 = pysam.TabixFile(annotation_file)\n",
    "# # read in the annotation file as a pandas dataframe\n",
    "# annot_df = pd.read_csv(annotation_file, sep='\\t', header=None, names=['chr', 'start', 'end', \n",
    "#                                                                       'attribute', 'feature', 'strand', \n",
    "#                                                                       'source', 'gene_name', 'gene_id', \n",
    "#                                                                       'gene_type', 'gene_level', 'exon_id', \n",
    "#                                                                       'exon_number', 'transcript_name',\n",
    "#                                                                       'transcript_type'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read the DMR sample file in to a pandas dataframe\n",
    "df = pd.read_table(DMR_file, low_memory=False, header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation file inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature_type_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12802</td>\n",
       "      <td>16450</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>Predicted enhancer region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13401</td>\n",
       "      <td>13600</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>CTCF binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14001</td>\n",
       "      <td>14200</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>CTCF binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15401</td>\n",
       "      <td>16600</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>CTCF binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25601</td>\n",
       "      <td>26600</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>CTCF binding site</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start    end       feature_type   feature_type_description\n",
       "0  chr1  12802  16450           Enhancer  Predicted enhancer region\n",
       "1  chr1  13401  13600  CTCF Binding Site          CTCF binding site\n",
       "2  chr1  14001  14200  CTCF Binding Site          CTCF binding site\n",
       "3  chr1  15401  16600  CTCF Binding Site          CTCF binding site\n",
       "4  chr1  25601  26600  CTCF Binding Site          CTCF binding site"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df_regulatory.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Enhancer', 'CTCF Binding Site', 'Promoter', 'Open chromatin',\n",
       "       'TF binding'], dtype=object)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop unwanted columns from annotation dfs\n",
    "# genes only:\n",
    "annot_df_gene = annot_df_gene.drop(['attribute'], axis=1)\n",
    "\n",
    "# all features:\n",
    "# annot_df = annot_df.drop(['attribute'], axis=1)\n",
    "\n",
    "# regulatory features:\n",
    "# what are the possible feature types in the regulatory build?\n",
    "annot_df_regulatory.feature_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # confirming that the gene annotation file matches the genes in all annotation file\n",
    "# # genes only:\n",
    "# display(annot_df_gene.loc[(annot_df_gene['chr']=='chr1')])\n",
    "# # all features:\n",
    "# display(annot_df.loc[(annot_df['chr']=='chr1')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DMR file inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure out syntax to add data type to silence warning --> must be specified dtype like np.int16 \n",
    "# see: https://stackoverflow.com/questions/24251219/pandas-read-csv-low-memory-and-dtype-options\n",
    "\n",
    "### Clean DMR file\n",
    "# remove CpGs from lenCpG values dataframe\n",
    "df['lenCpG'] = df['lenCpG'].apply(lambda x: x.split(\"CpGs\")[0])\n",
    "\n",
    "# remove bp from bp values in dataframe\n",
    "df['bp'] = df['bp'].apply(lambda x: x.split(\"bp\")[0])\n",
    "\n",
    "# rename chromosome column\n",
    "df.rename(columns = {'#chr':'chr'}, inplace = True)\n",
    "\n",
    "# uncomment below line if you want to drop unwanted columns\n",
    "#df = df.drop(['startCpG', 'endCpG', 'len_bp', 'region', 'tg_mean', 'bg_mean', 'delta_quants', 'delta_maxmin'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine if there is overlap between genes and DMRs\n",
    "# query and reference\n",
    "def overlap(q_st, q_end, ref_st, ref_end):\n",
    "    overlap  = min(q_end, ref_end)-max(q_st, ref_st)\n",
    "    return overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # loci is a row in a dataframe `a` with columns chr, start, end ['#chr', 'start', 'end']\n",
    "# # `tb` tabix is the indexed annotation file as a pysam.TabixFile object\n",
    "def gencode_annotate_loci(a, tb):\n",
    "    genes = []\n",
    "    gene_type = []\n",
    "    gene_level = []\n",
    "    gene_id = []\n",
    "    strand = []\n",
    "\n",
    "    try:\n",
    "        # reads the tabix file\n",
    "        # columns in tabix file: ['chr', 'start', 'end', 'attribute', 'feature', \n",
    "                                # 'strand', 'source', 'gene_name', 'gene_id','\n",
    "                                # gene_type', 'gene_level'])\n",
    "        \n",
    "        for region in tb.fetch(a['chr'], int(a['start']), int(a['end'])):\n",
    "            if region:\n",
    "                r = region.split('\\t')\n",
    "\n",
    "                ### Calculate length of the overlap between the query interval and the gencode gene\n",
    "                overlap_len = overlap(int(a['start']), int(a['end']), int(r[1]), int(r[2]))\n",
    "\n",
    "                ### first value is the percentage of the query interval that overlap with the gencode gene\n",
    "                ### second value is the percentage of the gencode gene that overlaps with the query interval\n",
    "                ret_val = '{}({})({})'.format(r[7], \n",
    "                                              np.round(overlap_len/float(int(a['end'])-int(a['start']))*100, 2), \n",
    "                                              np.round(overlap_len/float(int(r[2])-int(r[1]))*100, 2)\n",
    "                                              ) \n",
    "\n",
    "                genes.append(ret_val) \n",
    "                gene_type.append(r[9])\n",
    "                gene_level.append(r[10])\n",
    "                gene_id.append(r[8])\n",
    "                strand.append(r[5])\n",
    "    \n",
    "        if len(genes)>0:\n",
    "            return [\";\".join(genes), \";\".join(gene_type), \";\".join(gene_level), \";\".join(gene_id), \";\".join(strand)]\n",
    "        else:\n",
    "            return [\"NA(0)(0)\", \"NA\", np.nan, \"NA\", \"NA\"]\n",
    "    except ValueError:\n",
    "        return [\"NA(0)(0)\", \"NA\", np.nan, \"NA\", \"NA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to annotated DMRs with regulatory features and percent coverage\n",
    "# # loci is a row in a dataframe `a` with columns chr, start, end ['#chr', 'start', 'end']\n",
    "# # `tb` tabix is the indexed annotation file as a pysam.TabixFile object\n",
    "def regulatory_annotate_loci(a, tb):\n",
    "    reg_features = []\n",
    "\n",
    "    try:\n",
    "        # reads the tabix file\n",
    "        # columns in tabix file: ['chr', 'start', 'end', 'feature_type', 'feature_type_description']\n",
    "        \n",
    "        for region in tb.fetch(a['chr'], int(a['start']), int(a['end'])):\n",
    "            if region:\n",
    "                r = region.split('\\t')\n",
    "\n",
    "                ### Calculate length of the overlap between the query interval and the gencode gene\n",
    "                overlap_len = overlap(int(a['start']), int(a['end']), int(r[1]), int(r[2]))\n",
    "\n",
    "                ### first value is the percentage of the query interval that overlap with the gencode gene\n",
    "                ### second value is the percentage of the gencode gene that overlaps with the query interval\n",
    "                ret_val = '{}({})({})'.format(r[3], \n",
    "                                              np.round(overlap_len/float(int(a['end'])-int(a['start']))*100, 2), \n",
    "                                              np.round(overlap_len/float(int(r[2])-int(r[1]))*100, 2)\n",
    "                                              ) \n",
    "\n",
    "                reg_features.append(ret_val)\n",
    "    \n",
    "        if len(reg_features)>0:\n",
    "            return [\";\".join(reg_features)]\n",
    "        else:\n",
    "            return [\"NA(0)(0)\"]\n",
    "    except ValueError:\n",
    "        return [\"NA(0)(0)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to count the number of genes found for each region \n",
    "# by counting the number of semicolons in the gene_name or regulatory_feature column\n",
    "def count_features(gene):\n",
    "    if gene == 'NA(0)(0)':\n",
    "        return 0\n",
    "    else:\n",
    "        return gene.count(\";\") + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to return if region is intergenic or intragenic\n",
    "def genic(gene):\n",
    "    if gene == 'NA(0)(0)':\n",
    "        return \"intergenic\"\n",
    "    else:\n",
    "        return \"intragenic\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to return if region has regulatory features\n",
    "def regulatory(regulatory_feature):\n",
    "    if regulatory_feature == 'NA(0)(0)':\n",
    "        return \"no\"\n",
    "    else:\n",
    "        return \"yes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write a function to split the deliminates gene columns into lists\n",
    "def split_gene(gene_name, gene_type, gene_level, gene_id, strand):\n",
    "    gene_list = gene_name.split(\";\")\n",
    "    gt_list = str(gene_type).split(\";\") \n",
    "    gl_list = str(gene_level).split(\";\")\n",
    "    gid_list = str(gene_id).split(\";\")\n",
    "    strand_list = str(strand).split(\";\")\n",
    "    return gene_list, gt_list, gl_list, gid_list, strand_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_regulatory(regulatory_feature):\n",
    "    regulatory_list = regulatory_feature.split(\";\")\n",
    "    return regulatory_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate DMRs with gene information from gencode V44"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate dmrs using gencode_annotate_loci function and gencode_v44 tabix annotation file\n",
    "# returns genes, feature, gene_type, gene_level, exon_id, exon_number, transcript_name, transcript_type that overlap with DMRs\n",
    "# returns amount of overlap as a percentage under gene column as gene-name(overlap %)(overlap %)\n",
    "df[['gene_name', 'gene_type', 'gene_level', 'gene_id', 'strand']] = df.apply(lambda x: gencode_annotate_loci(x[['chr', 'start', 'end']], gencode_v44_genes), axis=1, result_type='expand')\n",
    "\n",
    "# replace empty strings with NaN\n",
    "df = df.replace(r'^\\s*$', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count number of genes in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene counts added to dataframe.\n",
      "Max number of genes in one region is 22.\n",
      "Regions with max number of genes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>delta_quants</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15177</th>\n",
       "      <td>chr5</td>\n",
       "      <td>141508878</td>\n",
       "      <td>141509045</td>\n",
       "      <td>8904372</td>\n",
       "      <td>8904378</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr5:141508878-141509045</td>\n",
       "      <td>6</td>\n",
       "      <td>167</td>\n",
       "      <td>0.661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.324</td>\n",
       "      <td>0.112</td>\n",
       "      <td>M</td>\n",
       "      <td>PCDHGA1(100.0)(0.09);PCDHGA2(100.0)(0.1);PCDHG...</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2</td>\n",
       "      <td>ENSG00000204956.6;ENSG00000081853.15;ENSG00000...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15178</th>\n",
       "      <td>chr5</td>\n",
       "      <td>141512826</td>\n",
       "      <td>141512943</td>\n",
       "      <td>8904452</td>\n",
       "      <td>8904458</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr5:141512826-141512943</td>\n",
       "      <td>6</td>\n",
       "      <td>117</td>\n",
       "      <td>0.592</td>\n",
       "      <td>...</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.309</td>\n",
       "      <td>0.999</td>\n",
       "      <td>M</td>\n",
       "      <td>PCDHGA1(100.0)(0.06);PCDHGA2(100.0)(0.07);PCDH...</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2</td>\n",
       "      <td>ENSG00000204956.6;ENSG00000081853.15;ENSG00000...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end  startCpG   endCpG        target  \\\n",
       "15177  chr5  141508878  141509045   8904372  8904378  colon_cancer   \n",
       "15178  chr5  141512826  141512943   8904452  8904458  colon_cancer   \n",
       "\n",
       "                         region lenCpG   bp  tg_mean  ...  delta_quants  \\\n",
       "15177  chr5:141508878-141509045      6  167    0.661  ...         0.344   \n",
       "15178  chr5:141512826-141512943      6  117    0.592  ...         0.374   \n",
       "\n",
       "       delta_maxmin  ttest  direction  \\\n",
       "15177         0.324  0.112          M   \n",
       "15178         0.309  0.999          M   \n",
       "\n",
       "                                               gene_name  \\\n",
       "15177  PCDHGA1(100.0)(0.09);PCDHGA2(100.0)(0.1);PCDHG...   \n",
       "15178  PCDHGA1(100.0)(0.06);PCDHGA2(100.0)(0.07);PCDH...   \n",
       "\n",
       "                                               gene_type  \\\n",
       "15177  protein_coding;protein_coding;protein_coding;p...   \n",
       "15178  protein_coding;protein_coding;protein_coding;p...   \n",
       "\n",
       "                                        gene_level  \\\n",
       "15177  2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2   \n",
       "15178  2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2   \n",
       "\n",
       "                                                 gene_id  \\\n",
       "15177  ENSG00000204956.6;ENSG00000081853.15;ENSG00000...   \n",
       "15178  ENSG00000204956.6;ENSG00000081853.15;ENSG00000...   \n",
       "\n",
       "                                            strand gene_num  \n",
       "15177  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+       22  \n",
       "15178  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+       22  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# count the number of genes found for each region and add to g_nums list\n",
    "g_nums = []\n",
    "for i,r in df.iterrows():\n",
    "    g_num = count_features(str(r['gene_name']))\n",
    "    g_nums.append(g_num)\n",
    "\n",
    "# add gene number column to dataframe\n",
    "df['gene_num'] = g_nums\n",
    "print(\"Gene counts added to dataframe.\")\n",
    "\n",
    "# return max number of genes found in one region and preview of regions with max number of genes\n",
    "max_g = max(g_nums)\n",
    "print(\"Max number of genes in one region is {}.\".format(max_g))\n",
    "print(\"Regions with max number of genes:\")\n",
    "display(df.loc[(df['gene_num']==max_g)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a variable for intergenic or intragenic classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply genic function to df\n",
    "df['genic'] = df['gene_name'].apply(genic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Regulatory feature annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate DMRs with regulatory information from ENSEMBL regulatory build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate dmrs using regulatory_annotate_loci function and regulatory build tabix annotation file\n",
    "# returns regulatory feature types that overlap with DMRs\n",
    "# returns amount of overlap as a percentage under regulatory_feature column as feature-name(overlap %)(overlap %)\n",
    "df[['regulatory_feature']] = df.apply(lambda x: regulatory_annotate_loci(x[['chr', 'start', 'end']], regulatory_build), axis=1, result_type='expand')\n",
    "\n",
    "# replace empty strings with NaN\n",
    "df = df.replace(r'^\\s*$', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count number of regulatory features in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Regulatory feature counts added to dataframe.\n",
      "Max number of regulatory features in one region is 5.\n",
      "Regions with max number of regulatory features:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>genic</th>\n",
       "      <th>regulatory_feature</th>\n",
       "      <th>reg_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39995</th>\n",
       "      <td>chr19</td>\n",
       "      <td>50115698</td>\n",
       "      <td>50117527</td>\n",
       "      <td>25827266</td>\n",
       "      <td>25827321</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr19:50115698-50117527</td>\n",
       "      <td>55</td>\n",
       "      <td>1829</td>\n",
       "      <td>0.531</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000287001(100.0)(3.49)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000287001.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>TF binding(24.99)(88.05);CTCF Binding Site(32....</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end  startCpG    endCpG        target  \\\n",
       "39995  chr19  50115698  50117527  25827266  25827321  colon_cancer   \n",
       "\n",
       "                        region lenCpG    bp  tg_mean  ...  direction  \\\n",
       "39995  chr19:50115698-50117527     55  1829    0.531  ...          U   \n",
       "\n",
       "                          gene_name  gene_type  gene_level            gene_id  \\\n",
       "39995  ENSG00000287001(100.0)(3.49)     lncRNA           2  ENSG00000287001.3   \n",
       "\n",
       "      strand gene_num       genic  \\\n",
       "39995      -        1  intragenic   \n",
       "\n",
       "                                      regulatory_feature reg_num  \n",
       "39995  TF binding(24.99)(88.05);CTCF Binding Site(32....       5  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# count the number of genes found for each region and add to g_nums list\n",
    "reg_nums = []\n",
    "for i,r in df.iterrows():\n",
    "    reg_num = count_features(str(r['regulatory_feature']))\n",
    "    reg_nums.append(reg_num)\n",
    "\n",
    "# add gene number column to dataframe\n",
    "df['reg_num'] = reg_nums\n",
    "print(\"Regulatory feature counts added to dataframe.\")\n",
    "\n",
    "# return max number of genes found in one region and preview of regions with max number of genes\n",
    "max_g = max(reg_nums)\n",
    "print(\"Max number of regulatory features in one region is {}.\".format(max_g))\n",
    "print(\"Regions with max number of regulatory features:\")\n",
    "display(df.loc[(df['reg_num']==max_g)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a variable indicating if there is a regulatory feature in the region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply regulatory function to df\n",
    "df['regulatory'] = df['regulatory_feature'].apply(regulatory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Observing DMR annotations and comparing to annotation files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>genic</th>\n",
       "      <th>regulatory_feature</th>\n",
       "      <th>reg_num</th>\n",
       "      <th>regulatory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Enhancer(1.38)(12.56)</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  \\\n",
       "0      6   416    0.311  ...   \n",
       "1      7   468    0.260  ...   \n",
       "2      8   206    0.633  ...   \n",
       "3     11  1811    0.236  ...   \n",
       "4     22  1169    0.435  ...   \n",
       "\n",
       "                                          gene_name  \\\n",
       "0                      ENSG00000238009(100.0)(0.94)   \n",
       "1                      ENSG00000238009(100.0)(1.05)   \n",
       "2  ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)   \n",
       "3                      ENSG00000228463(100.0)(1.78)   \n",
       "4                                          NA(0)(0)   \n",
       "\n",
       "                          gene_type  gene_level  \\\n",
       "0                            lncRNA           2   \n",
       "1                            lncRNA           2   \n",
       "2       lncRNA;processed_pseudogene         2;1   \n",
       "3  transcribed_processed_pseudogene           1   \n",
       "4                                NA         NaN   \n",
       "\n",
       "                               gene_id  strand gene_num       genic  \\\n",
       "0                    ENSG00000238009.6       -        1  intragenic   \n",
       "1                    ENSG00000238009.6       -        1  intragenic   \n",
       "2  ENSG00000238009.6;ENSG00000233750.3     -;+        2  intragenic   \n",
       "3                   ENSG00000228463.10       -        1  intragenic   \n",
       "4                                   NA      NA        0  intergenic   \n",
       "\n",
       "      regulatory_feature reg_num regulatory  \n",
       "0               NA(0)(0)       0         no  \n",
       "1               NA(0)(0)       0         no  \n",
       "2               NA(0)(0)       0         no  \n",
       "3  Enhancer(1.38)(12.56)       1        yes  \n",
       "4               NA(0)(0)       0         no  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()\n",
    "# annot_df_gene.head()\n",
    "# annot_df_regulatory.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking proper annotation (uncomment below lines to check different things)\n",
    "\n",
    "# # checking by chromosome\n",
    "# display(df.loc[(df['chr']=='chr4')])\n",
    "# display(annot_df_gene.loc[(annot_df_gene['chr']=='chr4')])\n",
    "# display(annot_df_regulatory.loc[(annot_df_gene['chr']=='chr4')])\n",
    "\n",
    "# # checking for specific genes\n",
    "# display(df.loc[(annot_df['gene_name']=='SEPTIN9') & (df['chr']=='chr17')])\n",
    "# display(annot_df_gene.loc[(annot_df_gene['gene_name']=='SEPTIN9') & (annot_df_gene['chr']=='chr17')])\n",
    "# display(annot_df_regulatory.loc[annot_df_regulatory['chr']=='chr17'])\n",
    "\n",
    "# # checking by chromosome and start/end\n",
    "# df_check = display(df.loc[(dmr['chr']=='chr17') & (df['start'].between(77280569,77500596))])\n",
    "# annot_check = display(annot_df_gene.loc[(annot_df_gene['chr']=='chr17') & (annot_df_gene['start'].between(77280569,77500596))])\n",
    "# annot_check_reg = display(annot_df_regulatory.loc[(annot_df_regulatory['chr']=='chr17') & (annot_df_regulatory['start'].between(77280569,77500596))])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean annotations "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split Gene names\n",
    "Splits gene names out for regions with more than one gene and creates new rows for df called `df_perGene`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split values in the columns gene, gene_type, gene_level by delimiter ; and add to a list of lists called new_rows\n",
    "new_rows = []\n",
    "for i,r in df.iterrows():\n",
    "    g_list, gt_list, gl_list, gid_list, strand_list = split_gene(r['gene_name'], r['gene_type'], r['gene_level'], r['gene_id'], r['strand'])\n",
    "\n",
    "    for g in range(len(g_list)):\n",
    "        new_rows.append(np.append(r[['chr', 'start', 'end', 'startCpG', \n",
    "                                     'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                     'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                     'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                     'direction', 'gene_name', 'gene_type',\n",
    "                                     'gene_level', 'gene_id', 'strand', 'genic', 'gene_num']\n",
    "                                     ].values, [g_list[g], gt_list[g], gl_list[g], \n",
    "                                                gid_list[g], strand_list[g]\n",
    "                                                ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check df against new rows to confirm proper splitting\n",
    "# df.head()\n",
    "# print(len(new_rows))\n",
    "# print(new_rows[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>gene_name_split</th>\n",
       "      <th>gene_type_split</th>\n",
       "      <th>gene_level_split</th>\n",
       "      <th>gene_id_split</th>\n",
       "      <th>strand_split</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27(100.0)(5.41)</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  gene_level                              gene_id  \\\n",
       "0      6   416    0.311  ...           2                    ENSG00000238009.6   \n",
       "1      7   468    0.260  ...           2                    ENSG00000238009.6   \n",
       "2      8   206    0.633  ...         2;1  ENSG00000238009.6;ENSG00000233750.3   \n",
       "3      8   206    0.633  ...         2;1  ENSG00000238009.6;ENSG00000233750.3   \n",
       "4     11  1811    0.236  ...           1                   ENSG00000228463.10   \n",
       "\n",
       "   strand       genic  gene_num               gene_name_split  \\\n",
       "0       -  intragenic         1  ENSG00000238009(100.0)(0.94)   \n",
       "1       -  intragenic         1  ENSG00000238009(100.0)(1.05)   \n",
       "2     -;+  intragenic         2  ENSG00000238009(100.0)(0.46)   \n",
       "3     -;+  intragenic         2           CICP27(100.0)(5.41)   \n",
       "4       -  intragenic         1  ENSG00000228463(100.0)(1.78)   \n",
       "\n",
       "                    gene_type_split gene_level_split       gene_id_split  \\\n",
       "0                            lncRNA                2   ENSG00000238009.6   \n",
       "1                            lncRNA                2   ENSG00000238009.6   \n",
       "2                            lncRNA                2   ENSG00000238009.6   \n",
       "3              processed_pseudogene                1   ENSG00000233750.3   \n",
       "4  transcribed_processed_pseudogene                1  ENSG00000228463.10   \n",
       "\n",
       "  strand_split  \n",
       "0            -  \n",
       "1            -  \n",
       "2            -  \n",
       "3            +  \n",
       "4            -  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add new rows to add to the dataframe\n",
    "df_perGene = pd.DataFrame(new_rows, columns=['chr', 'start', 'end', 'startCpG',\n",
    "                                            'endCpG', 'target', 'region', 'lenCpG',\n",
    "                                            'bp', 'tg_mean', 'bg_mean', 'delta_means',\n",
    "                                            'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                            'direction', 'gene_name', 'gene_type',\n",
    "                                            'gene_level', 'gene_id', 'strand', \n",
    "                                            'genic', 'gene_num', 'gene_name_split', \n",
    "                                            'gene_type_split', 'gene_level_split',\n",
    "                                            'gene_id_split', 'strand_split'])\n",
    "\n",
    "df_perGene.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split coverage percentages from gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>gene_name_split</th>\n",
       "      <th>gene_type_split</th>\n",
       "      <th>gene_level_split</th>\n",
       "      <th>gene_id_split</th>\n",
       "      <th>strand_split</th>\n",
       "      <th>gene_name_cleaned</th>\n",
       "      <th>region_coverage</th>\n",
       "      <th>gene_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27(100.0)(5.41)</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>100.0</td>\n",
       "      <td>5.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021(100.0)(0.08)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000235146(100.0)(2.7)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000235146.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235146</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>598963</td>\n",
       "      <td>599165</td>\n",
       "      <td>5342</td>\n",
       "      <td>5347</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:598963-599165</td>\n",
       "      <td>5</td>\n",
       "      <td>202</td>\n",
       "      <td>0.293</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000230021(100.0)(0.08)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>603632</td>\n",
       "      <td>604015</td>\n",
       "      <td>5406</td>\n",
       "      <td>5412</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:603632-604015</td>\n",
       "      <td>6</td>\n",
       "      <td>383</td>\n",
       "      <td>0.300</td>\n",
       "      <td>...</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000230021(100.0)(0.16)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "5  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "6  chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "7  chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "8  chr1  598963  599165      5342    5347  colon_cancer  chr1:598963-599165   \n",
       "9  chr1  603632  604015      5406    5412  colon_cancer  chr1:603632-604015   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...       genic  gene_num  \\\n",
       "0      6   416    0.311  ...  intragenic         1   \n",
       "1      7   468    0.260  ...  intragenic         1   \n",
       "2      8   206    0.633  ...  intragenic         2   \n",
       "3      8   206    0.633  ...  intragenic         2   \n",
       "4     11  1811    0.236  ...  intragenic         1   \n",
       "5     22  1169    0.435  ...  intergenic         0   \n",
       "6      7   193    0.122  ...  intragenic         2   \n",
       "7      7   193    0.122  ...  intragenic         2   \n",
       "8      5   202    0.293  ...  intragenic         1   \n",
       "9      6   383    0.300  ...  intragenic         1   \n",
       "\n",
       "                gene_name_split                   gene_type_split  \\\n",
       "0  ENSG00000238009(100.0)(0.94)                            lncRNA   \n",
       "1  ENSG00000238009(100.0)(1.05)                            lncRNA   \n",
       "2  ENSG00000238009(100.0)(0.46)                            lncRNA   \n",
       "3           CICP27(100.0)(5.41)              processed_pseudogene   \n",
       "4  ENSG00000228463(100.0)(1.78)  transcribed_processed_pseudogene   \n",
       "5                      NA(0)(0)                                NA   \n",
       "6  ENSG00000230021(100.0)(0.08)  transcribed_processed_pseudogene   \n",
       "7   ENSG00000235146(100.0)(2.7)                            lncRNA   \n",
       "8  ENSG00000230021(100.0)(0.08)  transcribed_processed_pseudogene   \n",
       "9  ENSG00000230021(100.0)(0.16)  transcribed_processed_pseudogene   \n",
       "\n",
       "   gene_level_split       gene_id_split strand_split gene_name_cleaned  \\\n",
       "0                 2   ENSG00000238009.6            -   ENSG00000238009   \n",
       "1                 2   ENSG00000238009.6            -   ENSG00000238009   \n",
       "2                 2   ENSG00000238009.6            -   ENSG00000238009   \n",
       "3                 1   ENSG00000233750.3            +            CICP27   \n",
       "4                 1  ENSG00000228463.10            -   ENSG00000228463   \n",
       "5               nan                  NA           NA                NA   \n",
       "6                 2  ENSG00000230021.10            -   ENSG00000230021   \n",
       "7                 2   ENSG00000235146.2            +   ENSG00000235146   \n",
       "8                 2  ENSG00000230021.10            -   ENSG00000230021   \n",
       "9                 2  ENSG00000230021.10            -   ENSG00000230021   \n",
       "\n",
       "  region_coverage gene_coverage  \n",
       "0           100.0          0.94  \n",
       "1           100.0          1.05  \n",
       "2           100.0          0.46  \n",
       "3           100.0          5.41  \n",
       "4           100.0          1.78  \n",
       "5               0             0  \n",
       "6           100.0          0.08  \n",
       "7           100.0           2.7  \n",
       "8           100.0          0.08  \n",
       "9           100.0          0.16  \n",
       "\n",
       "[10 rows x 31 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split gene_name column into gene_name, region_coverage, gene_coverage\n",
    "df_perGene['gene_name_cleaned'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[0].replace(\")\", \"\"))\n",
    "df_perGene['region_coverage'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[1].replace(\")\", \"\"))\n",
    "df_perGene['gene_coverage'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[2].replace(\")\", \"\"))\n",
    "\n",
    "df_perGene.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean and reorganize df_perGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>perc_gene_cov_region</th>\n",
       "      <th>perc_region_cov_gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.94</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>1.05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.46</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>5.41</td>\n",
       "      <td>100.0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>1.78</td>\n",
       "      <td>100.0</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  direction  num_genes        gene_name  \\\n",
       "0      6   416    0.311  ...          U          1  ENSG00000238009   \n",
       "1      7   468    0.260  ...          U          1  ENSG00000238009   \n",
       "2      8   206    0.633  ...          U          2  ENSG00000238009   \n",
       "3      8   206    0.633  ...          U          2           CICP27   \n",
       "4     11  1811    0.236  ...          U          1  ENSG00000228463   \n",
       "\n",
       "   perc_gene_cov_region  perc_region_cov_gene  \\\n",
       "0                  0.94                 100.0   \n",
       "1                  1.05                 100.0   \n",
       "2                  0.46                 100.0   \n",
       "3                  5.41                 100.0   \n",
       "4                  1.78                 100.0   \n",
       "\n",
       "                          gene_type  gene_level             gene_ID strand  \\\n",
       "0                            lncRNA           2   ENSG00000238009.6      -   \n",
       "1                            lncRNA           2   ENSG00000238009.6      -   \n",
       "2                            lncRNA           2   ENSG00000238009.6      -   \n",
       "3              processed_pseudogene           1   ENSG00000233750.3      +   \n",
       "4  transcribed_processed_pseudogene           1  ENSG00000228463.10      -   \n",
       "\n",
       "        genic  \n",
       "0  intragenic  \n",
       "1  intragenic  \n",
       "2  intragenic  \n",
       "3  intragenic  \n",
       "4  intragenic  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# once I have confirmed proper annotation, I can clean up the df\n",
    "# columns in df_per_gene: ['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
    "                    #    'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
    "                    #    'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
    "                    #    'gene_level', 'gene_id', 'strand', 'gene_num', 'gene_name_split',\n",
    "                    #    'gene_type_split', 'gene_level_split', 'gene_id_split', 'strand_split',\n",
    "                    #    'gene_name_cleaned', 'region_coverage', 'gene_coverage']\n",
    "\n",
    "## drop the genes column\n",
    "df_perGene = df_perGene.drop(['gene_name', 'gene_type','gene_level', \n",
    "                              'gene_id', 'strand', 'gene_name_split'], axis=1)\n",
    "\n",
    "# rename columns\n",
    "df_perGene.rename(columns = {'gene_name_cleaned':'gene_name', \n",
    "                              'gene_id_split':'gene_ID',\n",
    "                              'strand_split':'strand',\n",
    "                              'gene_type_split':'gene_type',\n",
    "                              'gene_level_split':'gene_level',\n",
    "                              'gene_num':'num_genes',\n",
    "                              'gene_coverage':'perc_gene_cov_region',\n",
    "                              'region_coverage':'perc_region_cov_gene'\n",
    "                              }, inplace = True)\n",
    "\n",
    "# reorder columns\n",
    "df_perGene = df_perGene.loc[:,['chr', 'start', 'end', 'startCpG', \n",
    "                                 'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                 'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                 'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                 'direction', 'num_genes', 'gene_name', \n",
    "                                 'perc_gene_cov_region', 'perc_region_cov_gene',\n",
    "                                 'gene_type', 'gene_level', 'gene_ID', 'strand', \n",
    "                                 'genic']]\n",
    "\n",
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split regulatory features\n",
    "\n",
    "Splits regulatory features out for regions with more than one regulatory feature and creates new rows for df called `df_perReg`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split values in the columns gene, gene_type, gene_level by delimiter ; and add to a list of lists called new_rows\n",
    "new_rows = []\n",
    "for i,r in df.iterrows():\n",
    "    reg_list = split_regulatory(r['regulatory_feature'])\n",
    "\n",
    "    for reg in range(len(reg_list)):\n",
    "        new_rows.append(np.append(r[['chr', 'start', 'end', 'startCpG', \n",
    "                                     'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                     'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                     'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                     'direction', 'gene_name', 'gene_type',\n",
    "                                     'gene_level', 'gene_id', 'strand', 'genic', \n",
    "                                     'gene_num','regulatory_feature', 'reg_num', \n",
    "                                     'regulatory']\n",
    "                                     ].values, [reg_list[reg]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # compare the df against new rows to confirm proper splitting\n",
    "# df.head()\n",
    "# print(len(new_rows))\n",
    "# print(new_rows[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add new rows to add to the dataframe (df_perReg)\n",
    "df_perReg = pd.DataFrame(new_rows, columns=['chr', 'start', 'end', 'startCpG', \n",
    "                                            'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                            'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                            'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                            'direction', 'gene_name', 'gene_type',\n",
    "                                            'gene_level', 'gene_id', 'strand', 'genic', \n",
    "                                            'gene_num','regulatory_feature', 'reg_num', \n",
    "                                            'regulatory', 'reg_split'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split coverage percentages from regulatory feature type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>regulatory_feature</th>\n",
       "      <th>reg_num</th>\n",
       "      <th>regulatory</th>\n",
       "      <th>reg_split</th>\n",
       "      <th>reg_name_cleaned</th>\n",
       "      <th>perc_reg_cov_region</th>\n",
       "      <th>perc_region_cov_reg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Enhancer(1.38)(12.56)</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Enhancer(1.38)(12.56)</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>1.38</td>\n",
       "      <td>12.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>598963</td>\n",
       "      <td>599165</td>\n",
       "      <td>5342</td>\n",
       "      <td>5347</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:598963-599165</td>\n",
       "      <td>5</td>\n",
       "      <td>202</td>\n",
       "      <td>0.293</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>603632</td>\n",
       "      <td>604015</td>\n",
       "      <td>5406</td>\n",
       "      <td>5412</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:603632-604015</td>\n",
       "      <td>6</td>\n",
       "      <td>383</td>\n",
       "      <td>0.300</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>605175</td>\n",
       "      <td>605246</td>\n",
       "      <td>5438</td>\n",
       "      <td>5446</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:605175-605246</td>\n",
       "      <td>8</td>\n",
       "      <td>71</td>\n",
       "      <td>0.454</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF Binding Site(35.21)(12.56)</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>CTCF Binding Site(35.21)(12.56)</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>35.21</td>\n",
       "      <td>12.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>605698</td>\n",
       "      <td>607036</td>\n",
       "      <td>5475</td>\n",
       "      <td>5494</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:605698-607036</td>\n",
       "      <td>19</td>\n",
       "      <td>1338</td>\n",
       "      <td>0.290</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621751</td>\n",
       "      <td>623283</td>\n",
       "      <td>5840</td>\n",
       "      <td>5857</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:621751-623283</td>\n",
       "      <td>17</td>\n",
       "      <td>1532</td>\n",
       "      <td>0.267</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>625843</td>\n",
       "      <td>626100</td>\n",
       "      <td>5869</td>\n",
       "      <td>5874</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:625843-626100</td>\n",
       "      <td>5</td>\n",
       "      <td>257</td>\n",
       "      <td>0.426</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631139</td>\n",
       "      <td>631224</td>\n",
       "      <td>5946</td>\n",
       "      <td>5951</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:631139-631224</td>\n",
       "      <td>5</td>\n",
       "      <td>85</td>\n",
       "      <td>0.066</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>Enhancer(100.0)(1.4);CTCF Binding Site(100.0)(...</td>\n",
       "      <td>2</td>\n",
       "      <td>yes</td>\n",
       "      <td>Enhancer(100.0)(1.4)</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631139</td>\n",
       "      <td>631224</td>\n",
       "      <td>5946</td>\n",
       "      <td>5951</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:631139-631224</td>\n",
       "      <td>5</td>\n",
       "      <td>85</td>\n",
       "      <td>0.066</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>Enhancer(100.0)(1.4);CTCF Binding Site(100.0)(...</td>\n",
       "      <td>2</td>\n",
       "      <td>yes</td>\n",
       "      <td>CTCF Binding Site(100.0)(14.19)</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>100.0</td>\n",
       "      <td>14.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631826</td>\n",
       "      <td>631979</td>\n",
       "      <td>5969</td>\n",
       "      <td>5974</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:631826-631979</td>\n",
       "      <td>5</td>\n",
       "      <td>153</td>\n",
       "      <td>0.202</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>Enhancer(100.0)(2.51);CTCF Binding Site(100.0)...</td>\n",
       "      <td>2</td>\n",
       "      <td>yes</td>\n",
       "      <td>Enhancer(100.0)(2.51)</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631826</td>\n",
       "      <td>631979</td>\n",
       "      <td>5969</td>\n",
       "      <td>5974</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:631826-631979</td>\n",
       "      <td>5</td>\n",
       "      <td>153</td>\n",
       "      <td>0.202</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>Enhancer(100.0)(2.51);CTCF Binding Site(100.0)...</td>\n",
       "      <td>2</td>\n",
       "      <td>yes</td>\n",
       "      <td>CTCF Binding Site(100.0)(25.54)</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>100.0</td>\n",
       "      <td>25.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>647626</td>\n",
       "      <td>649220</td>\n",
       "      <td>6156</td>\n",
       "      <td>6166</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:647626-649220</td>\n",
       "      <td>10</td>\n",
       "      <td>1594</td>\n",
       "      <td>0.167</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>658760</td>\n",
       "      <td>660493</td>\n",
       "      <td>6221</td>\n",
       "      <td>6240</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:658760-660493</td>\n",
       "      <td>19</td>\n",
       "      <td>1733</td>\n",
       "      <td>0.625</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>669492</td>\n",
       "      <td>669801</td>\n",
       "      <td>6298</td>\n",
       "      <td>6305</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:669492-669801</td>\n",
       "      <td>7</td>\n",
       "      <td>309</td>\n",
       "      <td>0.306</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685876</td>\n",
       "      <td>687330</td>\n",
       "      <td>6389</td>\n",
       "      <td>6399</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:685876-687330</td>\n",
       "      <td>10</td>\n",
       "      <td>1454</td>\n",
       "      <td>0.625</td>\n",
       "      <td>...</td>\n",
       "      <td>-;-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "      <td>CTCF Binding Site(13.69)(100.0);TF binding(40....</td>\n",
       "      <td>2</td>\n",
       "      <td>yes</td>\n",
       "      <td>CTCF Binding Site(13.69)(100.0)</td>\n",
       "      <td>CTCF Binding Site</td>\n",
       "      <td>13.69</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0   chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1   chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2   chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3   chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4   chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "5   chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "6   chr1  598963  599165      5342    5347  colon_cancer  chr1:598963-599165   \n",
       "7   chr1  603632  604015      5406    5412  colon_cancer  chr1:603632-604015   \n",
       "8   chr1  605175  605246      5438    5446  colon_cancer  chr1:605175-605246   \n",
       "9   chr1  605698  607036      5475    5494  colon_cancer  chr1:605698-607036   \n",
       "10  chr1  621751  623283      5840    5857  colon_cancer  chr1:621751-623283   \n",
       "11  chr1  625843  626100      5869    5874  colon_cancer  chr1:625843-626100   \n",
       "12  chr1  631139  631224      5946    5951  colon_cancer  chr1:631139-631224   \n",
       "13  chr1  631139  631224      5946    5951  colon_cancer  chr1:631139-631224   \n",
       "14  chr1  631826  631979      5969    5974  colon_cancer  chr1:631826-631979   \n",
       "15  chr1  631826  631979      5969    5974  colon_cancer  chr1:631826-631979   \n",
       "16  chr1  647626  649220      6156    6166  colon_cancer  chr1:647626-649220   \n",
       "17  chr1  658760  660493      6221    6240  colon_cancer  chr1:658760-660493   \n",
       "18  chr1  669492  669801      6298    6305  colon_cancer  chr1:669492-669801   \n",
       "19  chr1  685876  687330      6389    6399  colon_cancer  chr1:685876-687330   \n",
       "\n",
       "   lenCpG    bp  tg_mean  ...  strand       genic  gene_num  \\\n",
       "0       6   416    0.311  ...       -  intragenic         1   \n",
       "1       7   468    0.260  ...       -  intragenic         1   \n",
       "2       8   206    0.633  ...     -;+  intragenic         2   \n",
       "3      11  1811    0.236  ...       -  intragenic         1   \n",
       "4      22  1169    0.435  ...      NA  intergenic         0   \n",
       "5       7   193    0.122  ...     -;+  intragenic         2   \n",
       "6       5   202    0.293  ...       -  intragenic         1   \n",
       "7       6   383    0.300  ...       -  intragenic         1   \n",
       "8       8    71    0.454  ...       -  intragenic         1   \n",
       "9      19  1338    0.290  ...       -  intragenic         1   \n",
       "10     17  1532    0.267  ...       -  intragenic         1   \n",
       "11      5   257    0.426  ...       -  intragenic         1   \n",
       "12      5    85    0.066  ...     -;+  intragenic         2   \n",
       "13      5    85    0.066  ...     -;+  intragenic         2   \n",
       "14      5   153    0.202  ...     -;+  intragenic         2   \n",
       "15      5   153    0.202  ...     -;+  intragenic         2   \n",
       "16     10  1594    0.167  ...       -  intragenic         1   \n",
       "17     19  1733    0.625  ...       -  intragenic         1   \n",
       "18      7   309    0.306  ...       -  intragenic         1   \n",
       "19     10  1454    0.625  ...     -;-  intragenic         2   \n",
       "\n",
       "                                   regulatory_feature  reg_num regulatory  \\\n",
       "0                                            NA(0)(0)        0         no   \n",
       "1                                            NA(0)(0)        0         no   \n",
       "2                                            NA(0)(0)        0         no   \n",
       "3                               Enhancer(1.38)(12.56)        1        yes   \n",
       "4                                            NA(0)(0)        0         no   \n",
       "5                                            NA(0)(0)        0         no   \n",
       "6                                            NA(0)(0)        0         no   \n",
       "7                                            NA(0)(0)        0         no   \n",
       "8                     CTCF Binding Site(35.21)(12.56)        1        yes   \n",
       "9                                            NA(0)(0)        0         no   \n",
       "10                                           NA(0)(0)        0         no   \n",
       "11                                           NA(0)(0)        0         no   \n",
       "12  Enhancer(100.0)(1.4);CTCF Binding Site(100.0)(...        2        yes   \n",
       "13  Enhancer(100.0)(1.4);CTCF Binding Site(100.0)(...        2        yes   \n",
       "14  Enhancer(100.0)(2.51);CTCF Binding Site(100.0)...        2        yes   \n",
       "15  Enhancer(100.0)(2.51);CTCF Binding Site(100.0)...        2        yes   \n",
       "16                                           NA(0)(0)        0         no   \n",
       "17                                           NA(0)(0)        0         no   \n",
       "18                                           NA(0)(0)        0         no   \n",
       "19  CTCF Binding Site(13.69)(100.0);TF binding(40....        2        yes   \n",
       "\n",
       "                          reg_split   reg_name_cleaned perc_reg_cov_region  \\\n",
       "0                          NA(0)(0)                 NA                   0   \n",
       "1                          NA(0)(0)                 NA                   0   \n",
       "2                          NA(0)(0)                 NA                   0   \n",
       "3             Enhancer(1.38)(12.56)           Enhancer                1.38   \n",
       "4                          NA(0)(0)                 NA                   0   \n",
       "5                          NA(0)(0)                 NA                   0   \n",
       "6                          NA(0)(0)                 NA                   0   \n",
       "7                          NA(0)(0)                 NA                   0   \n",
       "8   CTCF Binding Site(35.21)(12.56)  CTCF Binding Site               35.21   \n",
       "9                          NA(0)(0)                 NA                   0   \n",
       "10                         NA(0)(0)                 NA                   0   \n",
       "11                         NA(0)(0)                 NA                   0   \n",
       "12             Enhancer(100.0)(1.4)           Enhancer               100.0   \n",
       "13  CTCF Binding Site(100.0)(14.19)  CTCF Binding Site               100.0   \n",
       "14            Enhancer(100.0)(2.51)           Enhancer               100.0   \n",
       "15  CTCF Binding Site(100.0)(25.54)  CTCF Binding Site               100.0   \n",
       "16                         NA(0)(0)                 NA                   0   \n",
       "17                         NA(0)(0)                 NA                   0   \n",
       "18                         NA(0)(0)                 NA                   0   \n",
       "19  CTCF Binding Site(13.69)(100.0)  CTCF Binding Site               13.69   \n",
       "\n",
       "   perc_region_cov_reg  \n",
       "0                    0  \n",
       "1                    0  \n",
       "2                    0  \n",
       "3                12.56  \n",
       "4                    0  \n",
       "5                    0  \n",
       "6                    0  \n",
       "7                    0  \n",
       "8                12.56  \n",
       "9                    0  \n",
       "10                   0  \n",
       "11                   0  \n",
       "12                 1.4  \n",
       "13               14.19  \n",
       "14                2.51  \n",
       "15               25.54  \n",
       "16                   0  \n",
       "17                   0  \n",
       "18                   0  \n",
       "19               100.0  \n",
       "\n",
       "[20 rows x 30 columns]"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split regulatory_feature column into reg feature type, region_coverage, feature_coverage\n",
    "df_perReg['reg_name_cleaned'] = df_perReg['reg_split'].apply(lambda x: x.split(\"(\")[0].replace(\")\", \"\"))\n",
    "df_perReg['perc_reg_cov_region'] = df_perReg['reg_split'].apply(lambda x: x.split(\"(\")[1].replace(\")\", \"\"))\n",
    "df_perReg['perc_region_cov_reg'] = df_perReg['reg_split'].apply(lambda x: x.split(\"(\")[2].replace(\")\", \"\"))\n",
    "\n",
    "df_perReg.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
       "       'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
       "       'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
       "       'gene_level', 'gene_id', 'strand', 'genic', 'gene_num',\n",
       "       'regulatory_feature', 'reg_num', 'regulatory', 'reg_split',\n",
       "       'reg_name_cleaned', 'perc_reg_cov_region', 'perc_region_cov_reg'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_perReg.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean and reorganize df_perReg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>direction</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>perc_gene_cov_region</th>\n",
       "      <th>perc_region_cov_gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.94</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>1.05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.46</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>5.41</td>\n",
       "      <td>100.0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>1.78</td>\n",
       "      <td>100.0</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  direction  num_genes        gene_name  \\\n",
       "0      6   416    0.311  ...          U          1  ENSG00000238009   \n",
       "1      7   468    0.260  ...          U          1  ENSG00000238009   \n",
       "2      8   206    0.633  ...          U          2  ENSG00000238009   \n",
       "3      8   206    0.633  ...          U          2           CICP27   \n",
       "4     11  1811    0.236  ...          U          1  ENSG00000228463   \n",
       "\n",
       "   perc_gene_cov_region  perc_region_cov_gene  \\\n",
       "0                  0.94                 100.0   \n",
       "1                  1.05                 100.0   \n",
       "2                  0.46                 100.0   \n",
       "3                  5.41                 100.0   \n",
       "4                  1.78                 100.0   \n",
       "\n",
       "                          gene_type  gene_level             gene_ID strand  \\\n",
       "0                            lncRNA           2   ENSG00000238009.6      -   \n",
       "1                            lncRNA           2   ENSG00000238009.6      -   \n",
       "2                            lncRNA           2   ENSG00000238009.6      -   \n",
       "3              processed_pseudogene           1   ENSG00000233750.3      +   \n",
       "4  transcribed_processed_pseudogene           1  ENSG00000228463.10      -   \n",
       "\n",
       "        genic  \n",
       "0  intragenic  \n",
       "1  intragenic  \n",
       "2  intragenic  \n",
       "3  intragenic  \n",
       "4  intragenic  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# once I have confirmed proper annotation, I can clean up the df\n",
    "# columns in df_perReg:  ['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
    "                    #    'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
    "                    #    'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
    "                    #    'gene_level', 'gene_id', 'strand', 'genic', 'gene_num',\n",
    "                    #    'regulatory_feature', 'reg_num', 'regulatory', 'reg_split',\n",
    "                    #    'reg_name_cleaned', 'perc_reg_cov_region', 'perc_region_cov_reg']\n",
    "\n",
    "## drop the genes column\n",
    "df_perReg = df_perReg.drop(['regulatory_feature','reg_split'], axis=1)\n",
    "\n",
    "# rename columns\n",
    "df_perReg.rename(columns = {'reg_name_cleaned':'regulatory_feature',\n",
    "                             'reg_num':'num_regs',\n",
    "                             'gene_num':'num_genes'\n",
    "                             }, inplace = True)\n",
    "\n",
    "# reorder columns\n",
    "df_perReg = df_perReg.loc[:,['chr', 'start', 'end', 'startCpG', \n",
    "                             'endCpG', 'target', 'region', 'lenCpG', \n",
    "                             'bp', 'tg_mean', 'bg_mean', 'delta_means',\n",
    "                             'delta_quants', 'delta_maxmin', 'ttest',\n",
    "                             'direction', 'num_regs', 'regulatory_feature', \n",
    "                             'perc_reg_cov_region', 'perc_region_cov_reg', \n",
    "                             'num_genes', 'gene_name', 'gene_type', \n",
    "                             'gene_level', 'gene_id', 'strand', 'genic', \n",
    "                             'regulatory']]\n",
    "\n",
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save files\n",
    "\n",
    "df has columns: ['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
    "                'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
    "                'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
    "                'gene_level', 'gene_id', 'strand', 'gene_num', 'genic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv with all regions (partial annotation/no dups) has been created\n",
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_per_gene.tsv with all regions (dup regions to split genes) has been created\n"
     ]
    }
   ],
   "source": [
    "df_perGene.to_csv(output_file_per_gene, index=False, header=True, sep=\"\\t\")\n",
    "df_perReg.to_csv(output_file_per_reg, index=False, header=True, sep=\"\\t\")\n",
    "df.to_csv(output_file, index=False, header=True, sep=\"\\t\")\n",
    "\n",
    "print(output_file + \" with all regions (partial annotation/no dups) has been created\")\n",
    "print(output_file_per_gene + \" with all regions (dup regions to split genes) has been created\")\n",
    "print(output_file_per_reg + \" with all regions (dup regions to split regulatory features) has been created\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Misc. code chunks from dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # display dmr dataframe where gene_num is greater than 1\n",
    "# df.loc[(df['gene_num'] > 1)]\n",
    "\n",
    "# # return rows where feature is exon in df_perGene\n",
    "# df_perGene[df_perGene.feature=='exon'].tail(25)\n",
    "\n",
    "## skip if you want to keep regions with no genes\n",
    "## Transforms the data frame and removes anything that returned NA\n",
    "## Remove all the intervals that do not overlap with genes\n",
    "# df = df[df['genes'] != \"NA(0)\"].reset_index(drop=True)\n",
    "\n",
    "# # display items in list of lists\n",
    "# print(new_rows[0:5])\n",
    "\n",
    "# # listing column names\n",
    "# column_names = list(df.columns.values)\n",
    "# print(column_names)\n",
    "\n",
    "# # exporting top 100 regions with lowest p-values for annotation in browser (USC Genome Browser has gencode V44 annotation)\n",
    "# df.sort_values(by=['ttest'], inplace=True)\n",
    "# dmr_exp = df.loc[(df['ttest']<0.050)].head(100)\n",
    "# dmr_exp = dmr_exp.drop(['startCpG', 'endCpG', 'target', 'bp',\n",
    "#                    'lenCpG', 'region', 'tg_mean', 'bg_mean', \n",
    "#                    'delta_means', 'delta_quants', 'delta_maxmin',\n",
    "#                    'ttest', 'direction'], axis=1).head(100)\n",
    "# dmr_exp.to_csv('top100_colon_markers.tsv', sep=\"\\t\", header=False, index=False) \n",
    "# dmr_exp.head()\n",
    "\n",
    "# # create dataframe with only significant DMRs\n",
    "# dmr = df_perGene.loc[(df_perGene['ttest'] < 0.05)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions for keller\n",
    "\n",
    "What do the various chromosome labels mean? \n",
    "Are these contigs that are not included in chromosome contructed references?\n",
    "    Does `M` refer to mitrochondrial?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GL000009.2', 'GL000194.1', 'GL000195.1', 'GL000205.2',\n",
       "       'GL000209.2', 'GL000213.1', 'GL000216.2', 'GL000218.1',\n",
       "       'GL000219.1', 'GL000220.1', 'GL000225.1', 'GL000250.2',\n",
       "       'GL000251.2', 'GL000252.2', 'GL000253.2', 'GL000254.2',\n",
       "       'GL000255.2', 'GL000256.2', 'GL000257.2', 'GL000258.2',\n",
       "       'GL339449.2', 'GL383518.1', 'GL383519.1', 'GL383520.2',\n",
       "       'GL383521.1', 'GL383522.1', 'GL383526.1', 'GL383527.1',\n",
       "       'GL383528.1', 'GL383531.1', 'GL383532.1', 'GL383533.1',\n",
       "       'GL383534.2', 'GL383539.1', 'GL383540.1', 'GL383541.1',\n",
       "       'GL383542.1', 'GL383545.1', 'GL383546.1', 'GL383550.2',\n",
       "       'GL383551.1', 'GL383552.1', 'GL383553.2', 'GL383554.1',\n",
       "       'GL383555.2', 'GL383556.1', 'GL383557.1', 'GL383563.3',\n",
       "       'GL383564.2', 'GL383565.1', 'GL383566.1', 'GL383567.1',\n",
       "       'GL383569.1', 'GL383570.1', 'GL383571.1', 'GL383572.1',\n",
       "       'GL383573.1', 'GL383574.1', 'GL383575.2', 'GL383576.1',\n",
       "       'GL383577.2', 'GL383579.2', 'GL383580.2', 'GL383581.2',\n",
       "       'GL383582.2', 'GL383583.2', 'GL582966.2', 'GL877875.1',\n",
       "       'GL877876.1', 'GL949742.1', 'GL949746.1', 'GL949747.2',\n",
       "       'GL949748.2', 'GL949749.2', 'GL949750.2', 'GL949751.2',\n",
       "       'GL949752.1', 'GL949753.2', 'JH159136.1', 'JH159137.1',\n",
       "       'JH159146.1', 'JH159147.1', 'JH159148.1', 'JH636055.2',\n",
       "       'KB021644.2', 'KB663609.1', 'KI270442.1', 'KI270711.1',\n",
       "       'KI270713.1', 'KI270721.1', 'KI270726.1', 'KI270727.1',\n",
       "       'KI270728.1', 'KI270731.1', 'KI270733.1', 'KI270734.1',\n",
       "       'KI270744.1', 'KI270750.1', 'KI270758.1', 'KI270759.1',\n",
       "       'KI270760.1', 'KI270761.1', 'KI270762.1', 'KI270763.1',\n",
       "       'KI270765.1', 'KI270766.1', 'KI270767.1', 'KI270768.1',\n",
       "       'KI270769.1', 'KI270770.1', 'KI270772.1', 'KI270773.1',\n",
       "       'KI270774.1', 'KI270776.1', 'KI270777.1', 'KI270779.1',\n",
       "       'KI270780.1', 'KI270781.1', 'KI270782.1', 'KI270783.1',\n",
       "       'KI270784.1', 'KI270785.1', 'KI270786.1', 'KI270788.1',\n",
       "       'KI270789.1', 'KI270790.1', 'KI270791.1', 'KI270792.1',\n",
       "       'KI270793.1', 'KI270794.1', 'KI270795.1', 'KI270796.1',\n",
       "       'KI270797.1', 'KI270798.1', 'KI270799.1', 'KI270800.1',\n",
       "       'KI270801.1', 'KI270802.1', 'KI270803.1', 'KI270804.1',\n",
       "       'KI270805.1', 'KI270806.1', 'KI270807.1', 'KI270808.1',\n",
       "       'KI270809.1', 'KI270810.1', 'KI270811.1', 'KI270812.1',\n",
       "       'KI270813.1', 'KI270814.1', 'KI270815.1', 'KI270816.1',\n",
       "       'KI270817.1', 'KI270818.1', 'KI270819.1', 'KI270821.1',\n",
       "       'KI270822.1', 'KI270823.1', 'KI270824.1', 'KI270825.1',\n",
       "       'KI270829.1', 'KI270830.1', 'KI270831.1', 'KI270832.1',\n",
       "       'KI270833.1', 'KI270834.1', 'KI270835.1', 'KI270837.1',\n",
       "       'KI270838.1', 'KI270840.1', 'KI270842.1', 'KI270844.1',\n",
       "       'KI270845.1', 'KI270846.1', 'KI270847.1', 'KI270848.1',\n",
       "       'KI270849.1', 'KI270850.1', 'KI270851.1', 'KI270852.1',\n",
       "       'KI270853.1', 'KI270854.1', 'KI270855.1', 'KI270856.1',\n",
       "       'KI270857.1', 'KI270858.1', 'KI270859.1', 'KI270860.1',\n",
       "       'KI270861.1', 'KI270862.1', 'KI270863.1', 'KI270865.1',\n",
       "       'KI270866.1', 'KI270867.1', 'KI270868.1', 'KI270869.1',\n",
       "       'KI270870.1', 'KI270871.1', 'KI270872.1', 'KI270873.1',\n",
       "       'KI270874.1', 'KI270875.1', 'KI270876.1', 'KI270877.1',\n",
       "       'KI270878.1', 'KI270879.1', 'KI270880.1', 'KI270881.1',\n",
       "       'KI270882.1', 'KI270883.1', 'KI270884.1', 'KI270885.1',\n",
       "       'KI270886.1', 'KI270887.1', 'KI270888.1', 'KI270889.1',\n",
       "       'KI270890.1', 'KI270891.1', 'KI270892.1', 'KI270893.1',\n",
       "       'KI270894.1', 'KI270895.1', 'KI270896.1', 'KI270897.1',\n",
       "       'KI270898.1', 'KI270899.1', 'KI270900.1', 'KI270901.1',\n",
       "       'KI270902.1', 'KI270903.1', 'KI270904.1', 'KI270905.1',\n",
       "       'KI270906.1', 'KI270907.1', 'KI270908.1', 'KI270909.1',\n",
       "       'KI270910.1', 'KI270911.1', 'KI270912.1', 'KI270913.1',\n",
       "       'KI270914.1', 'KI270915.1', 'KI270916.1', 'KI270917.1',\n",
       "       'KI270918.1', 'KI270919.1', 'KI270920.1', 'KI270921.1',\n",
       "       'KI270922.1', 'KI270923.1', 'KI270924.1', 'KI270925.1',\n",
       "       'KI270926.1', 'KI270927.1', 'KI270928.1', 'KI270929.1',\n",
       "       'KI270930.1', 'KI270931.1', 'KI270932.1', 'KI270933.1',\n",
       "       'KI270934.1', 'KI270935.1', 'KI270936.1', 'KI270937.1',\n",
       "       'KI270938.1', 'KN196472.1', 'KN196473.1', 'KN196474.1',\n",
       "       'KN196475.1', 'KN196476.1', 'KN196477.1', 'KN196478.1',\n",
       "       'KN196479.1', 'KN196480.1', 'KN196481.1', 'KN196484.1',\n",
       "       'KN196485.1', 'KN196486.1', 'KN196487.1', 'KN538360.1',\n",
       "       'KN538361.1', 'KN538362.1', 'KN538363.1', 'KN538364.1',\n",
       "       'KN538368.1', 'KN538369.1', 'KN538370.1', 'KN538371.1',\n",
       "       'KN538372.1', 'KN538373.1', 'KQ031383.1', 'KQ031384.1',\n",
       "       'KQ031385.1', 'KQ031387.1', 'KQ031388.1', 'KQ031389.1',\n",
       "       'KQ090014.1', 'KQ090015.1', 'KQ090016.1', 'KQ090018.1',\n",
       "       'KQ090021.1', 'KQ090022.1', 'KQ090026.1', 'KQ090027.1',\n",
       "       'KQ090028.1', 'KQ458382.1', 'KQ458383.1', 'KQ458384.1',\n",
       "       'KQ458385.1', 'KQ458386.1', 'KQ458387.1', 'KQ458388.1',\n",
       "       'KQ759759.2', 'KQ759760.1', 'KQ759761.1', 'KQ759762.2',\n",
       "       'KQ983255.1', 'KQ983256.1', 'KQ983257.1', 'KQ983258.1',\n",
       "       'KV575243.1', 'KV575244.1', 'KV575245.1', 'KV575246.1',\n",
       "       'KV575247.1', 'KV575248.1', 'KV575249.1', 'KV575250.1',\n",
       "       'KV575251.1', 'KV575252.1', 'KV575253.1', 'KV575254.1',\n",
       "       'KV575255.1', 'KV575256.1', 'KV575257.1', 'KV575258.1',\n",
       "       'KV575259.1', 'KV575260.1', 'KV766192.1', 'KV766193.1',\n",
       "       'KV766194.1', 'KV766195.1', 'KV766196.1', 'KV766198.1',\n",
       "       'KV880763.1', 'KV880764.1', 'KV880765.1', 'KV880766.1',\n",
       "       'KV880768.1', 'KZ208904.1', 'KZ208905.1', 'KZ208906.1',\n",
       "       'KZ208907.1', 'KZ208908.1', 'KZ208909.1', 'KZ208911.1',\n",
       "       'KZ208912.1', 'KZ208913.1', 'KZ208914.1', 'KZ208915.1',\n",
       "       'KZ208916.1', 'KZ208917.1', 'KZ208919.1', 'KZ208920.1',\n",
       "       'KZ208921.1', 'KZ208922.1', 'KZ208924.1', 'KZ559102.1',\n",
       "       'KZ559103.1', 'KZ559104.1', 'KZ559105.1', 'KZ559106.1',\n",
       "       'KZ559107.1', 'KZ559108.1', 'KZ559109.1', 'KZ559110.1',\n",
       "       'KZ559111.1', 'KZ559112.1', 'KZ559113.1', 'KZ559114.1',\n",
       "       'KZ559115.1', 'KZ559116.1', 'ML143341.1', 'ML143342.1',\n",
       "       'ML143343.1', 'ML143344.1', 'ML143345.1', 'ML143347.1',\n",
       "       'ML143349.1', 'ML143350.1', 'ML143352.1', 'ML143354.1',\n",
       "       'ML143355.1', 'ML143357.1', 'ML143358.1', 'ML143359.1',\n",
       "       'ML143360.1', 'ML143361.1', 'ML143362.1', 'ML143365.1',\n",
       "       'ML143366.1', 'ML143367.1', 'ML143369.1', 'ML143370.1',\n",
       "       'ML143371.1', 'ML143372.1', 'ML143373.1', 'ML143374.1',\n",
       "       'ML143375.1', 'ML143376.1', 'ML143377.1', 'ML143378.1',\n",
       "       'ML143380.1', 'ML143381.1', 'MU273330.1', 'MU273331.1',\n",
       "       'MU273332.1', 'MU273333.1', 'MU273334.1', 'MU273335.1',\n",
       "       'MU273336.1', 'MU273337.1', 'MU273338.1', 'MU273339.1',\n",
       "       'MU273340.1', 'MU273341.1', 'MU273342.1', 'MU273343.1',\n",
       "       'MU273344.1', 'MU273345.1', 'MU273346.1', 'MU273347.1',\n",
       "       'MU273348.1', 'MU273349.1', 'MU273351.1', 'MU273352.1',\n",
       "       'MU273353.1', 'MU273355.1', 'MU273356.1', 'MU273357.1',\n",
       "       'MU273358.1', 'MU273359.1', 'MU273360.1', 'MU273361.1',\n",
       "       'MU273362.1', 'MU273363.1', 'MU273364.1', 'MU273365.1',\n",
       "       'MU273366.1', 'MU273367.1', 'MU273368.1', 'MU273369.1',\n",
       "       'MU273370.1', 'MU273371.1', 'MU273372.1', 'MU273373.1',\n",
       "       'MU273374.1', 'MU273375.1', 'MU273376.1', 'MU273377.1',\n",
       "       'MU273378.1', 'MU273379.1', 'MU273380.1', 'MU273381.1',\n",
       "       'MU273382.1', 'MU273383.1', 'MU273384.1', 'MU273385.1',\n",
       "       'MU273387.1', 'MU273388.1', 'MU273389.1', 'MU273390.1',\n",
       "       'MU273391.1', 'MU273392.1', 'MU273393.1', 'MU273394.1',\n",
       "       'MU273395.1', 'MU273396.1', 'MU273397.1', 'MU273398.1', 'chr1',\n",
       "       'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16',\n",
       "       'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21', 'chr22',\n",
       "       'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrM',\n",
       "       'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# figure out how to interpret the chromosome labels\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=\n",
    "annot_df.chr.unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "annot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
