{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conda environment to use: annot\n",
    "\n",
    "# References for sources used to write this script:\n",
    "#   - Gene annotation methods: https://medium.com/intothegenomics/annotate-genes-and-genomic-coordinates-using-python-9259efa6ffc2\n",
    "#   - Gencode annotation file format: https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "# figure out how to interpret the chromosome labels\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=\n",
    "\n",
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "\n",
    "# Assign directories and files to use for annotation reference, DMRs, and output\n",
    "annotation_file = \"/Users/meghansleeper/Desktop/farm-files/dmr-workflow/supplementary/annotation_beds/gencode.v44.annotation.all.sorted.formatted.bed.gz\"\n",
    "annotation_file_genes = \"/Users/meghansleeper/Desktop/farm-files/dmr-workflow/supplementary/annotation_beds/gencode.v44.annotation.genes.sorted.formatted.bed.gz\"\n",
    "DMR_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers.tsv\"\n",
    "output_file_per_gene = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_per_gene.tsv\"\n",
    "output_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv\"\n",
    "dmr_output_file = \"/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_dmrs.tsv\"\n",
    "\n",
    "# Read reference annotation file with tabix\n",
    "gencode_v44 = pysam.TabixFile(annotation_file)\n",
    "gencode_v44_genes = pysam.TabixFile(annotation_file_genes)\n",
    "\n",
    "# read annotation files into dataframes\n",
    "annot_df = pd.read_csv(annotation_file, sep='\\t', header=None, names=['chr', 'start', 'end', \n",
    "                                                                      'attribute', 'feature', 'strand', \n",
    "                                                                      'source', 'gene_name', 'gene_id', \n",
    "                                                                      'gene_type', 'gene_level', 'exon_id', \n",
    "                                                                      'exon_number', 'transcript_name',\n",
    "                                                                      'transcript_type'])\n",
    "\n",
    "\n",
    "annot_df_gene = pd.read_csv(annotation_file_genes, sep='\\t', header=None, names=['chr', 'start', 'end', \n",
    "                                                                                'attribute', 'feature', 'strand', \n",
    "                                                                                'source', 'gene_name', 'gene_id', \n",
    "                                                                                'gene_type', 'gene_level'])\n",
    "\n",
    "### Read the sample file in to a pandas dataframe\n",
    "df = pd.read_table(DMR_file, low_memory=False, header=0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation file inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unwanted columns from annotation dfs\n",
    "annot_df_gene = annot_df_gene.drop(['attribute'], axis=1)\n",
    "annot_df = annot_df.drop(['attribute'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196077</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196078</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196079</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196080</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196081</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388929</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918364</td>\n",
       "      <td>248919146</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001312561.6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>PGBD2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388930</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918364</td>\n",
       "      <td>248919946</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001400348.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>PGBD2-202</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388931</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388932</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPL23AP25-201</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388933</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001666859.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL23AP25-201</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192857 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end          feature strand  source  gene_name  \\\n",
       "196077  chr1      11869      12227             exon      +  HAVANA    DDX11L2   \n",
       "196078  chr1      11869      14409             gene      +  HAVANA    DDX11L2   \n",
       "196079  chr1      11869      14409       transcript      +  HAVANA    DDX11L2   \n",
       "196080  chr1      12010      12057             exon      +  HAVANA    DDX11L1   \n",
       "196081  chr1      12010      13670             gene      +  HAVANA    DDX11L1   \n",
       "...      ...        ...        ...              ...    ...     ...        ...   \n",
       "388929  chr1  248918364  248919146  three_prime_UTR      +  HAVANA      PGBD2   \n",
       "388930  chr1  248918364  248919946  three_prime_UTR      +  HAVANA      PGBD2   \n",
       "388931  chr1  248936581  248937043             gene      +  HAVANA  RPL23AP25   \n",
       "388932  chr1  248936581  248937043       transcript      +  HAVANA  RPL23AP25   \n",
       "388933  chr1  248936581  248937043             exon      +  HAVANA  RPL23AP25   \n",
       "\n",
       "                   gene_id                           gene_type  gene_level  \\\n",
       "196077   ENSG00000290825.1                              lncRNA           2   \n",
       "196078   ENSG00000290825.1                              lncRNA           2   \n",
       "196079   ENSG00000290825.1                              lncRNA           2   \n",
       "196080   ENSG00000223972.6  transcribed_unprocessed_pseudogene           2   \n",
       "196081   ENSG00000223972.6  transcribed_unprocessed_pseudogene           2   \n",
       "...                    ...                                 ...         ...   \n",
       "388929  ENSG00000185220.12                      protein_coding           2   \n",
       "388930  ENSG00000185220.12                      protein_coding           2   \n",
       "388931   ENSG00000233084.2                processed_pseudogene           1   \n",
       "388932   ENSG00000233084.2                processed_pseudogene           1   \n",
       "388933   ENSG00000233084.2                processed_pseudogene           1   \n",
       "\n",
       "                  exon_id  exon_number transcript_name  \\\n",
       "196077  ENSE00002234944.1          1.0     DDX11L2-202   \n",
       "196078                NaN          NaN             NaN   \n",
       "196079                NaN          NaN     DDX11L2-202   \n",
       "196080  ENSE00001948541.1          1.0     DDX11L1-201   \n",
       "196081                NaN          NaN             NaN   \n",
       "...                   ...          ...             ...   \n",
       "388929  ENSE00001312561.6          3.0       PGBD2-201   \n",
       "388930  ENSE00001400348.4          3.0       PGBD2-202   \n",
       "388931                NaN          NaN             NaN   \n",
       "388932                NaN          NaN   RPL23AP25-201   \n",
       "388933  ENSE00001666859.1          1.0   RPL23AP25-201   \n",
       "\n",
       "                           transcript_type  \n",
       "196077                              lncRNA  \n",
       "196078                                 NaN  \n",
       "196079                              lncRNA  \n",
       "196080  transcribed_unprocessed_pseudogene  \n",
       "196081                                 NaN  \n",
       "...                                    ...  \n",
       "388929                      protein_coding  \n",
       "388930                      protein_coding  \n",
       "388931                                 NaN  \n",
       "388932                processed_pseudogene  \n",
       "388933                processed_pseudogene  \n",
       "\n",
       "[192857 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7416</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7417</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7418</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7419</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7420</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13070</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248850006</td>\n",
       "      <td>248859144</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13071</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248859164</td>\n",
       "      <td>248864796</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ZNF692-DT</td>\n",
       "      <td>ENSG00000227237.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13072</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13073</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248912690</td>\n",
       "      <td>248912795</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>RNU6-1205P</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13074</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5659 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end feature strand   source    gene_name  \\\n",
       "7416   chr1      11869      14409    gene      +   HAVANA      DDX11L2   \n",
       "7417   chr1      12010      13670    gene      +   HAVANA      DDX11L1   \n",
       "7418   chr1      14404      29570    gene      -   HAVANA       WASH7P   \n",
       "7419   chr1      17369      17436    gene      -  ENSEMBL    MIR6859-1   \n",
       "7420   chr1      29554      31109    gene      +   HAVANA  MIR1302-2HG   \n",
       "...     ...        ...        ...     ...    ...      ...          ...   \n",
       "13070  chr1  248850006  248859144    gene      -   HAVANA       ZNF692   \n",
       "13071  chr1  248859164  248864796    gene      +   HAVANA    ZNF692-DT   \n",
       "13072  chr1  248906196  248919946    gene      +   HAVANA        PGBD2   \n",
       "13073  chr1  248912690  248912795    gene      -  ENSEMBL   RNU6-1205P   \n",
       "13074  chr1  248936581  248937043    gene      +   HAVANA    RPL23AP25   \n",
       "\n",
       "                  gene_id                           gene_type  gene_level  \n",
       "7416    ENSG00000290825.1                              lncRNA           2  \n",
       "7417    ENSG00000223972.6  transcribed_unprocessed_pseudogene           2  \n",
       "7418    ENSG00000227232.5              unprocessed_pseudogene           2  \n",
       "7419    ENSG00000278267.1                               miRNA           3  \n",
       "7420    ENSG00000243485.5                              lncRNA           2  \n",
       "...                   ...                                 ...         ...  \n",
       "13070  ENSG00000171163.16                      protein_coding           1  \n",
       "13071   ENSG00000227237.1                              lncRNA           1  \n",
       "13072  ENSG00000185220.12                      protein_coding           2  \n",
       "13073   ENSG00000200495.1                               snRNA           3  \n",
       "13074   ENSG00000233084.2                processed_pseudogene           1  \n",
       "\n",
       "[5659 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# # confirming that the gene annotation file matches the genes in all annotation file\n",
    "display(annot_df.loc[(annot_df['chr']=='chr1')])\n",
    "display(annot_df_gene.loc[(annot_df_gene['chr']=='chr1')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DMR file inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure out syntax to add data type to silence warning --> must be specified dtype like np.int16 \n",
    "# see: https://stackoverflow.com/questions/24251219/pandas-read-csv-low-memory-and-dtype-options\n",
    "\n",
    "### Clean DMR file\n",
    "# remove CpGs from lenCpG values dataframe\n",
    "df['lenCpG'] = df['lenCpG'].apply(lambda x: x.split(\"CpGs\")[0])\n",
    "\n",
    "# remove bp from bp values in dataframe\n",
    "df['bp'] = df['bp'].apply(lambda x: x.split(\"bp\")[0])\n",
    "\n",
    "# rename chromosome column\n",
    "df.rename(columns = {'#chr':'chr'}, inplace = True)\n",
    "\n",
    "# uncomment below line if you want to drop unwanted columns\n",
    "#df = df.drop(['startCpG', 'endCpG', 'len_bp', 'region', 'tg_mean', 'bg_mean', 'delta_quants', 'delta_maxmin'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine if there is overlap between genes and DMRs\n",
    "# query and reference\n",
    "def overlap(q_st, q_end, ref_st, ref_end):\n",
    "    overlap  = min(q_end, ref_end)-max(q_st, ref_st)\n",
    "    return overlap\n",
    "\n",
    "\n",
    "# # loci is a row in a dataframe `a` with columns chr, start, end ['#chr', 'start', 'end']\n",
    "# # `tb` tabix is the indexed annotation file as a pysam.TabixFile object\n",
    "def gencode_annotate_loci(a, tb):\n",
    "    genes = []\n",
    "    gene_type = []\n",
    "    gene_level = []\n",
    "    gene_id = []\n",
    "    strand = []\n",
    "\n",
    "    try:\n",
    "        # reads the tabix file\n",
    "        # columns in tabix file: ['chr', 'start', 'end', 'attribute', 'feature', \n",
    "                                # 'strand', 'source', 'gene_name', 'gene_id','\n",
    "                                # gene_type', 'gene_level'])\n",
    "        \n",
    "        for region in tb.fetch(a['chr'], int(a['start']), int(a['end'])):\n",
    "            if region:\n",
    "                r = region.split('\\t')\n",
    "\n",
    "                ### Calculate length of the overlap between the query interval and the gencode gene\n",
    "                overlap_len = overlap(int(a['start']), int(a['end']), int(r[1]), int(r[2]))\n",
    "\n",
    "                ### first value is the percentage of the query interval that overlap with the gencode gene\n",
    "                ### second value is the percentage of the gencode gene that overlaps with the query interval\n",
    "                ret_val = '{}({})({})'.format(r[7], \n",
    "                                              np.round(overlap_len/float(int(a['end'])-int(a['start']))*100, 2), \n",
    "                                              np.round(overlap_len/float(int(r[2])-int(r[1]))*100, 2)\n",
    "                                              ) \n",
    "\n",
    "                genes.append(ret_val) \n",
    "                gene_type.append(r[9])\n",
    "                gene_level.append(r[10])\n",
    "                gene_id.append(r[8])\n",
    "                strand.append(r[5])\n",
    "    \n",
    "        if len(genes)>0:\n",
    "            return [\";\".join(genes), \";\".join(gene_type), \";\".join(gene_level), \";\".join(gene_id), \";\".join(strand)]\n",
    "        else:\n",
    "            return [\"NA(0)(0)\", \"NA\", np.nan, \"NA\", \"NA\"]\n",
    "    except ValueError:\n",
    "        return [\"NA(0)(0)\", \"NA\", np.nan, \"NA\", \"NA\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate DMRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate dmrs using gencode_annotate_loci function and gencode_v44 annotation file\n",
    "# returns genes, feature, gene_type, gene_level, exon_id, exon_number, transcript_name, transcript_type that overlap with DMRs\n",
    "# returns amount of overlap as a percentage under gene column as gene-name(overlap %)\n",
    "df[['gene_name', 'gene_type', 'gene_level', 'gene_id', 'strand']] = df.apply(lambda x: gencode_annotate_loci(x[['chr', 'start', 'end']], gencode_v44_genes), axis=1, result_type='expand')\n",
    "\n",
    "# replace empty strings with NaN\n",
    "df = df.replace(r'^\\s*$', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Observing DMR annotations and comparing to annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>delta_means</th>\n",
       "      <th>delta_quants</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.413</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.356</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  delta_means  delta_quants  delta_maxmin   ttest  \\\n",
       "0      6   416    0.311  ...        0.413         0.372         0.330  0.1260   \n",
       "1      7   468    0.260  ...        0.407         0.360         0.314  0.1460   \n",
       "2      8   206    0.633  ...        0.367         0.356         0.344  0.0761   \n",
       "3     11  1811    0.236  ...        0.389         0.382         0.375  0.1640   \n",
       "4     22  1169    0.435  ...        0.440         0.407         0.375  0.0927   \n",
       "\n",
       "   direction                                         gene_name  \\\n",
       "0          U                      ENSG00000238009(100.0)(0.94)   \n",
       "1          U                      ENSG00000238009(100.0)(1.05)   \n",
       "2          U  ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)   \n",
       "3          U                      ENSG00000228463(100.0)(1.78)   \n",
       "4          U                                          NA(0)(0)   \n",
       "\n",
       "                          gene_type gene_level  \\\n",
       "0                            lncRNA          2   \n",
       "1                            lncRNA          2   \n",
       "2       lncRNA;processed_pseudogene        2;1   \n",
       "3  transcribed_processed_pseudogene          1   \n",
       "4                                NA        NaN   \n",
       "\n",
       "                               gene_id strand  \n",
       "0                    ENSG00000238009.6      -  \n",
       "1                    ENSG00000238009.6      -  \n",
       "2  ENSG00000238009.6;ENSG00000233750.3    -;+  \n",
       "3                   ENSG00000228463.10      -  \n",
       "4                                   NA     NA  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df_gene.head()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking proper annotation (uncomment below lines to check different things)\n",
    "\n",
    "# # checking by chromosome\n",
    "# display(df.loc[(df['chr']=='chr4')])\n",
    "# display(annot_df_gene.loc[(annot_df_gene['chr']=='chr4')])\n",
    "\n",
    "# # checking for specific genes\n",
    "# display(df.loc[(annot_df['gene_name']=='SEPTIN9') & (df['chr']=='chr17')])\n",
    "# display(annot_df.loc[(annot_df['gene_name']=='SEPTIN9') & (annot_df['chr']=='chr17')])\n",
    "\n",
    "# # checking by chromosome and start/end\n",
    "# df_check = display(df.loc[(dmr['chr']=='chr17') & (df['start'].between(77280569,77500596))])\n",
    "# annot_check = display(annot_df.loc[(annot_df['chr']=='chr17') & (annot_df['start'].between(77280569,77500596))])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a variable for intergenic or intragenic classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to return if region is intergenic or intragenic\n",
    "def genic(gene):\n",
    "    if gene == 'NA(0)(0)':\n",
    "        return \"intergenic\"\n",
    "    else:\n",
    "        return \"intragenic\"\n",
    "    \n",
    "# apply genic function to df\n",
    "df['genic'] = df['gene_name'].apply(genic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count genes per region and create gene_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to count the number of genes found for each region \n",
    "# by counting the number of semicolons in the gene_name column\n",
    "def count_genes(gene):\n",
    "    if gene == 'NA(0)(0)':\n",
    "        return 0\n",
    "    else:\n",
    "        return gene.count(\";\") + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene counts added to dataframe.\n",
      "Max number of genes in one region is 22.\n",
      "Regions with max number of genes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15177</th>\n",
       "      <td>chr5</td>\n",
       "      <td>141508878</td>\n",
       "      <td>141509045</td>\n",
       "      <td>8904372</td>\n",
       "      <td>8904378</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr5:141508878-141509045</td>\n",
       "      <td>6</td>\n",
       "      <td>167</td>\n",
       "      <td>0.661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.324</td>\n",
       "      <td>0.112</td>\n",
       "      <td>M</td>\n",
       "      <td>PCDHGA1(100.0)(0.09);PCDHGA2(100.0)(0.1);PCDHG...</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2</td>\n",
       "      <td>ENSG00000204956.6;ENSG00000081853.15;ENSG00000...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15178</th>\n",
       "      <td>chr5</td>\n",
       "      <td>141512826</td>\n",
       "      <td>141512943</td>\n",
       "      <td>8904452</td>\n",
       "      <td>8904458</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr5:141512826-141512943</td>\n",
       "      <td>6</td>\n",
       "      <td>117</td>\n",
       "      <td>0.592</td>\n",
       "      <td>...</td>\n",
       "      <td>0.309</td>\n",
       "      <td>0.999</td>\n",
       "      <td>M</td>\n",
       "      <td>PCDHGA1(100.0)(0.06);PCDHGA2(100.0)(0.07);PCDH...</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2</td>\n",
       "      <td>ENSG00000204956.6;ENSG00000081853.15;ENSG00000...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end  startCpG   endCpG        target  \\\n",
       "15177  chr5  141508878  141509045   8904372  8904378  colon_cancer   \n",
       "15178  chr5  141512826  141512943   8904452  8904458  colon_cancer   \n",
       "\n",
       "                         region lenCpG   bp  tg_mean  ...  delta_maxmin  \\\n",
       "15177  chr5:141508878-141509045      6  167    0.661  ...         0.324   \n",
       "15178  chr5:141512826-141512943      6  117    0.592  ...         0.309   \n",
       "\n",
       "       ttest  direction                                          gene_name  \\\n",
       "15177  0.112          M  PCDHGA1(100.0)(0.09);PCDHGA2(100.0)(0.1);PCDHG...   \n",
       "15178  0.999          M  PCDHGA1(100.0)(0.06);PCDHGA2(100.0)(0.07);PCDH...   \n",
       "\n",
       "                                               gene_type  \\\n",
       "15177  protein_coding;protein_coding;protein_coding;p...   \n",
       "15178  protein_coding;protein_coding;protein_coding;p...   \n",
       "\n",
       "                                        gene_level  \\\n",
       "15177  2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2   \n",
       "15178  2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2;2   \n",
       "\n",
       "                                                 gene_id  \\\n",
       "15177  ENSG00000204956.6;ENSG00000081853.15;ENSG00000...   \n",
       "15178  ENSG00000204956.6;ENSG00000081853.15;ENSG00000...   \n",
       "\n",
       "                                            strand       genic gene_num  \n",
       "15177  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+  intragenic       22  \n",
       "15178  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+  intragenic       22  \n",
       "\n",
       "[2 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# count the number of genes found for each region and add to g_nums list\n",
    "g_nums = []\n",
    "for i,r in df.iterrows():\n",
    "    g_num = count_genes(str(r['gene_name']))\n",
    "    g_nums.append(g_num)\n",
    "\n",
    "# add gene number column to dataframe\n",
    "df['gene_num'] = g_nums\n",
    "print(\"Gene counts added to dataframe.\")\n",
    "\n",
    "# return max number of genes found in one region and preview of regions with max number of genes\n",
    "max_g = max(g_nums)\n",
    "print(\"Max number of genes in one region is {}.\".format(max_g))\n",
    "print(\"Regions with max number of genes:\")\n",
    "display(df.loc[(df['gene_num']==max_g)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split gene names out for regions with more than one gene\n",
    "\n",
    "This section creates new rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write a function to split the deliminates gene columns into lists\n",
    "def split_gene(gene_name, gene_type, gene_level, gene_id, strand):\n",
    "    gene_list = gene_name.split(\";\")\n",
    "    gt_list = str(gene_type).split(\";\") \n",
    "    gl_list = str(gene_level).split(\";\")\n",
    "    gid_list = str(gene_id).split(\";\")\n",
    "    strand_list = str(strand).split(\";\")\n",
    "    return gene_list, gt_list, gl_list, gid_list, strand_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split values in the columns gene, gene_type, gene_level by delimiter ; and add to a list of lists called new_rows\n",
    "new_rows = []\n",
    "for i,r in df.iterrows():\n",
    "    g_list, gt_list, gl_list, gid_list, strand_list = split_gene(r['gene_name'], r['gene_type'], r['gene_level'], r['gene_id'], r['strand'])\n",
    "\n",
    "    for g in range(len(g_list)):\n",
    "        new_rows.append(np.append(r[['chr', 'start', 'end', 'startCpG', \n",
    "                                     'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                     'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                     'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                     'direction', 'gene_name', 'gene_type',\n",
    "                                     'gene_level', 'gene_id', 'strand', 'gene_num']\n",
    "                                     ].values, [g_list[g], gt_list[g], gl_list[g], \n",
    "                                                gid_list[g], strand_list[g]\n",
    "                                                ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  delta_maxmin   ttest  direction  \\\n",
       "0      6   416    0.311  ...         0.330  0.1260          U   \n",
       "1      7   468    0.260  ...         0.314  0.1460          U   \n",
       "2      8   206    0.633  ...         0.344  0.0761          U   \n",
       "3     11  1811    0.236  ...         0.375  0.1640          U   \n",
       "4     22  1169    0.435  ...         0.375  0.0927          U   \n",
       "\n",
       "                                          gene_name  \\\n",
       "0                      ENSG00000238009(100.0)(0.94)   \n",
       "1                      ENSG00000238009(100.0)(1.05)   \n",
       "2  ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)   \n",
       "3                      ENSG00000228463(100.0)(1.78)   \n",
       "4                                          NA(0)(0)   \n",
       "\n",
       "                          gene_type gene_level  \\\n",
       "0                            lncRNA          2   \n",
       "1                            lncRNA          2   \n",
       "2       lncRNA;processed_pseudogene        2;1   \n",
       "3  transcribed_processed_pseudogene          1   \n",
       "4                                NA        NaN   \n",
       "\n",
       "                               gene_id strand       genic gene_num  \n",
       "0                    ENSG00000238009.6      -  intragenic        1  \n",
       "1                    ENSG00000238009.6      -  intragenic        1  \n",
       "2  ENSG00000238009.6;ENSG00000233750.3    -;+  intragenic        2  \n",
       "3                   ENSG00000228463.10      -  intragenic        1  \n",
       "4                                   NA     NA  intergenic        0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50317\n",
      "[array(['chr1', 92977, 93393, 1045, 1051, 'colon_cancer',\n",
      "       'chr1:92977-93393', '6', '416', 0.311, 0.724, 0.413, 0.372, 0.33,\n",
      "       0.126, 'U', 'ENSG00000238009(100.0)(0.94)', 'lncRNA', '2',\n",
      "       'ENSG00000238009.6', '-', 1, 'ENSG00000238009(100.0)(0.94)',\n",
      "       'lncRNA', '2', 'ENSG00000238009.6', '-'], dtype=object), array(['chr1', 99252, 99720, 1094, 1101, 'colon_cancer',\n",
      "       'chr1:99252-99720', '7', '468', 0.26, 0.667, 0.407, 0.36, 0.314,\n",
      "       0.146, 'U', 'ENSG00000238009(100.0)(1.05)', 'lncRNA', '2',\n",
      "       'ENSG00000238009.6', '-', 1, 'ENSG00000238009(100.0)(1.05)',\n",
      "       'lncRNA', '2', 'ENSG00000238009.6', '-'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U', 'ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)',\n",
      "       'lncRNA;processed_pseudogene', '2;1',\n",
      "       'ENSG00000238009.6;ENSG00000233750.3', '-;+', 2,\n",
      "       'ENSG00000238009(100.0)(0.46)', 'lncRNA', '2', 'ENSG00000238009.6',\n",
      "       '-'], dtype=object), array(['chr1', 132975, 133181, 1336, 1344, 'colon_cancer',\n",
      "       'chr1:132975-133181', '8', '206', 0.633, 1.0, 0.367, 0.356, 0.344,\n",
      "       0.0761, 'U', 'ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)',\n",
      "       'lncRNA;processed_pseudogene', '2;1',\n",
      "       'ENSG00000238009.6;ENSG00000233750.3', '-;+', 2,\n",
      "       'CICP27(100.0)(5.41)', 'processed_pseudogene', '1',\n",
      "       'ENSG00000233750.3', '+'], dtype=object), array(['chr1', 275815, 277626, 2790, 2801, 'colon_cancer',\n",
      "       'chr1:275815-277626', '11', '1811', 0.236, 0.625, 0.389, 0.382,\n",
      "       0.375, 0.164, 'U', 'ENSG00000228463(100.0)(1.78)',\n",
      "       'transcribed_processed_pseudogene', '1', 'ENSG00000228463.10', '-',\n",
      "       1, 'ENSG00000228463(100.0)(1.78)',\n",
      "       'transcribed_processed_pseudogene', '1', 'ENSG00000228463.10', '-'],\n",
      "      dtype=object)]\n"
     ]
    }
   ],
   "source": [
    "print(len(new_rows))\n",
    "print(new_rows[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add new rows to add to the dataframe\n",
    "df_perGene = pd.DataFrame(new_rows, columns=['chr', 'start', 'end', 'startCpG',\n",
    "                                            'endCpG', 'target', 'region', 'lenCpG',\n",
    "                                            'bp', 'tg_mean', 'bg_mean', 'delta_means',\n",
    "                                            'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                            'direction', 'gene_name', 'gene_type',\n",
    "                                            'gene_level', 'gene_id', 'strand', \n",
    "                                            'gene_num', 'gene_name_split', \n",
    "                                            'gene_type_split', 'gene_level_split',\n",
    "                                            'gene_id_split', 'strand_split'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>gene_name_split</th>\n",
       "      <th>gene_type_split</th>\n",
       "      <th>gene_level_split</th>\n",
       "      <th>gene_id_split</th>\n",
       "      <th>strand_split</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27(100.0)(5.41)</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...                         gene_type  gene_level  \\\n",
       "0      6   416    0.311  ...                            lncRNA           2   \n",
       "1      7   468    0.260  ...                            lncRNA           2   \n",
       "2      8   206    0.633  ...       lncRNA;processed_pseudogene         2;1   \n",
       "3      8   206    0.633  ...       lncRNA;processed_pseudogene         2;1   \n",
       "4     11  1811    0.236  ...  transcribed_processed_pseudogene           1   \n",
       "\n",
       "                               gene_id  strand  gene_num  \\\n",
       "0                    ENSG00000238009.6       -         1   \n",
       "1                    ENSG00000238009.6       -         1   \n",
       "2  ENSG00000238009.6;ENSG00000233750.3     -;+         2   \n",
       "3  ENSG00000238009.6;ENSG00000233750.3     -;+         2   \n",
       "4                   ENSG00000228463.10       -         1   \n",
       "\n",
       "                gene_name_split                   gene_type_split  \\\n",
       "0  ENSG00000238009(100.0)(0.94)                            lncRNA   \n",
       "1  ENSG00000238009(100.0)(1.05)                            lncRNA   \n",
       "2  ENSG00000238009(100.0)(0.46)                            lncRNA   \n",
       "3           CICP27(100.0)(5.41)              processed_pseudogene   \n",
       "4  ENSG00000228463(100.0)(1.78)  transcribed_processed_pseudogene   \n",
       "\n",
       "  gene_level_split       gene_id_split strand_split  \n",
       "0                2   ENSG00000238009.6            -  \n",
       "1                2   ENSG00000238009.6            -  \n",
       "2                2   ENSG00000238009.6            -  \n",
       "3                1   ENSG00000233750.3            +  \n",
       "4                1  ENSG00000228463.10            -  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split out coverage percentage from gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>gene_name_split</th>\n",
       "      <th>gene_type_split</th>\n",
       "      <th>gene_level_split</th>\n",
       "      <th>gene_id_split</th>\n",
       "      <th>strand_split</th>\n",
       "      <th>gene_name_cleaned</th>\n",
       "      <th>region_coverage</th>\n",
       "      <th>gene_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27(100.0)(5.41)</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>100.0</td>\n",
       "      <td>5.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021(100.0)(0.08)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588665</td>\n",
       "      <td>588858</td>\n",
       "      <td>5055</td>\n",
       "      <td>5062</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:588665-588858</td>\n",
       "      <td>7</td>\n",
       "      <td>193</td>\n",
       "      <td>0.122</td>\n",
       "      <td>...</td>\n",
       "      <td>-;+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000235146(100.0)(2.7)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000235146.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000235146</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>598963</td>\n",
       "      <td>599165</td>\n",
       "      <td>5342</td>\n",
       "      <td>5347</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:598963-599165</td>\n",
       "      <td>5</td>\n",
       "      <td>202</td>\n",
       "      <td>0.293</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000230021(100.0)(0.08)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>603632</td>\n",
       "      <td>604015</td>\n",
       "      <td>5406</td>\n",
       "      <td>5412</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:603632-604015</td>\n",
       "      <td>6</td>\n",
       "      <td>383</td>\n",
       "      <td>0.300</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000230021(100.0)(0.16)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "5  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "6  chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "7  chr1  588665  588858      5055    5062  colon_cancer  chr1:588665-588858   \n",
       "8  chr1  598963  599165      5342    5347  colon_cancer  chr1:598963-599165   \n",
       "9  chr1  603632  604015      5406    5412  colon_cancer  chr1:603632-604015   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  strand  gene_num               gene_name_split  \\\n",
       "0      6   416    0.311  ...       -         1  ENSG00000238009(100.0)(0.94)   \n",
       "1      7   468    0.260  ...       -         1  ENSG00000238009(100.0)(1.05)   \n",
       "2      8   206    0.633  ...     -;+         2  ENSG00000238009(100.0)(0.46)   \n",
       "3      8   206    0.633  ...     -;+         2           CICP27(100.0)(5.41)   \n",
       "4     11  1811    0.236  ...       -         1  ENSG00000228463(100.0)(1.78)   \n",
       "5     22  1169    0.435  ...      NA         0                      NA(0)(0)   \n",
       "6      7   193    0.122  ...     -;+         2  ENSG00000230021(100.0)(0.08)   \n",
       "7      7   193    0.122  ...     -;+         2   ENSG00000235146(100.0)(2.7)   \n",
       "8      5   202    0.293  ...       -         1  ENSG00000230021(100.0)(0.08)   \n",
       "9      6   383    0.300  ...       -         1  ENSG00000230021(100.0)(0.16)   \n",
       "\n",
       "                    gene_type_split  gene_level_split       gene_id_split  \\\n",
       "0                            lncRNA                 2   ENSG00000238009.6   \n",
       "1                            lncRNA                 2   ENSG00000238009.6   \n",
       "2                            lncRNA                 2   ENSG00000238009.6   \n",
       "3              processed_pseudogene                 1   ENSG00000233750.3   \n",
       "4  transcribed_processed_pseudogene                 1  ENSG00000228463.10   \n",
       "5                                NA               nan                  NA   \n",
       "6  transcribed_processed_pseudogene                 2  ENSG00000230021.10   \n",
       "7                            lncRNA                 2   ENSG00000235146.2   \n",
       "8  transcribed_processed_pseudogene                 2  ENSG00000230021.10   \n",
       "9  transcribed_processed_pseudogene                 2  ENSG00000230021.10   \n",
       "\n",
       "  strand_split gene_name_cleaned region_coverage gene_coverage  \n",
       "0            -   ENSG00000238009           100.0          0.94  \n",
       "1            -   ENSG00000238009           100.0          1.05  \n",
       "2            -   ENSG00000238009           100.0          0.46  \n",
       "3            +            CICP27           100.0          5.41  \n",
       "4            -   ENSG00000228463           100.0          1.78  \n",
       "5           NA                NA               0             0  \n",
       "6            -   ENSG00000230021           100.0          0.08  \n",
       "7            +   ENSG00000235146           100.0           2.7  \n",
       "8            -   ENSG00000230021           100.0          0.08  \n",
       "9            -   ENSG00000230021           100.0          0.16  \n",
       "\n",
       "[10 rows x 30 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split gene_name column into gene_name, region_coverage, gene_coverage\n",
    "df_perGene['gene_name_cleaned'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[0].replace(\")\", \"\"))\n",
    "df_perGene['region_coverage'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[1].replace(\")\", \"\"))\n",
    "df_perGene['gene_coverage'] = df_perGene['gene_name_split'].apply(lambda x: x.split(\"(\")[2].replace(\")\", \"\"))\n",
    "\n",
    "df_perGene.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
       "       'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
       "       'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
       "       'gene_level', 'gene_id', 'strand', 'gene_num', 'gene_name_split',\n",
       "       'gene_type_split', 'gene_level_split', 'gene_id_split', 'strand_split',\n",
       "       'gene_name_cleaned', 'region_coverage', 'gene_coverage'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_perGene.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean and reorganize columns if df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>prop_gene_cov_region</th>\n",
       "      <th>prop_region_cov_gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.94</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>1.05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>0.46</td>\n",
       "      <td>100.0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>2</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>5.41</td>\n",
       "      <td>100.0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>1.78</td>\n",
       "      <td>100.0</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "4  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...   ttest  direction  num_genes        gene_name  \\\n",
       "0      6   416    0.311  ...  0.1260          U          1  ENSG00000238009   \n",
       "1      7   468    0.260  ...  0.1460          U          1  ENSG00000238009   \n",
       "2      8   206    0.633  ...  0.0761          U          2  ENSG00000238009   \n",
       "3      8   206    0.633  ...  0.0761          U          2           CICP27   \n",
       "4     11  1811    0.236  ...  0.1640          U          1  ENSG00000228463   \n",
       "\n",
       "   prop_gene_cov_region prop_region_cov_gene  \\\n",
       "0                  0.94                100.0   \n",
       "1                  1.05                100.0   \n",
       "2                  0.46                100.0   \n",
       "3                  5.41                100.0   \n",
       "4                  1.78                100.0   \n",
       "\n",
       "                          gene_type gene_level             gene_ID strand  \n",
       "0                            lncRNA          2   ENSG00000238009.6      -  \n",
       "1                            lncRNA          2   ENSG00000238009.6      -  \n",
       "2                            lncRNA          2   ENSG00000238009.6      -  \n",
       "3              processed_pseudogene          1   ENSG00000233750.3      +  \n",
       "4  transcribed_processed_pseudogene          1  ENSG00000228463.10      -  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# once I have confirmed proper annotation, I can clean up the df\n",
    "# columns in df_per_gene: ['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
    "                    #    'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
    "                    #    'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
    "                    #    'gene_level', 'gene_id', 'strand', 'gene_num', 'gene_name_split',\n",
    "                    #    'gene_type_split', 'gene_level_split', 'gene_id_split', 'strand_split',\n",
    "                    #    'gene_name_cleaned', 'region_coverage', 'gene_coverage']\n",
    "\n",
    "## drop the genes column\n",
    "df_perGene = df_perGene.drop(['gene_name', 'gene_type','gene_level', \n",
    "                              'gene_id', 'strand', 'gene_name_split'], axis=1)\n",
    "\n",
    "# rename columns\n",
    "df_perGene.rename(columns = {'gene_name_cleaned':'gene_name', \n",
    "                              'gene_id_split':'gene_ID',\n",
    "                              'strand_split':'strand',\n",
    "                              'gene_type_split':'gene_type',\n",
    "                              'gene_level_split':'gene_level',\n",
    "                              'gene_num':'num_genes',\n",
    "                              'gene_coverage':'prop_gene_cov_region',\n",
    "                              'region_coverage':'prop_region_cov_gene'\n",
    "                              }, inplace = True)\n",
    "\n",
    "# reorder columns\n",
    "df_perGene = df_perGene.loc[:,['chr', 'start', 'end', 'startCpG', \n",
    "                                 'endCpG', 'target', 'region', 'lenCpG', \n",
    "                                 'bp', 'tg_mean', 'bg_mean', 'delta_means', \n",
    "                                 'delta_quants', 'delta_maxmin', 'ttest', \n",
    "                                 'direction', 'num_genes', 'gene_name', \n",
    "                                 'prop_gene_cov_region', 'prop_region_cov_gene',\n",
    "                                 'gene_type', 'gene_level', 'gene_ID', 'strand']]\n",
    "\n",
    "df_perGene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe with only significant DMRs\n",
    "dmr = df_perGene.loc[(df_perGene['ttest'] < 0.05)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>startCpG</th>\n",
       "      <th>endCpG</th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>lenCpG</th>\n",
       "      <th>bp</th>\n",
       "      <th>tg_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>delta_maxmin</th>\n",
       "      <th>ttest</th>\n",
       "      <th>direction</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>genic</th>\n",
       "      <th>gene_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92977</td>\n",
       "      <td>93393</td>\n",
       "      <td>1045</td>\n",
       "      <td>1051</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:92977-93393</td>\n",
       "      <td>6</td>\n",
       "      <td>416</td>\n",
       "      <td>0.311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.94)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99252</td>\n",
       "      <td>99720</td>\n",
       "      <td>1094</td>\n",
       "      <td>1101</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:99252-99720</td>\n",
       "      <td>7</td>\n",
       "      <td>468</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(1.05)</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132975</td>\n",
       "      <td>133181</td>\n",
       "      <td>1336</td>\n",
       "      <td>1344</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:132975-133181</td>\n",
       "      <td>8</td>\n",
       "      <td>206</td>\n",
       "      <td>0.633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344</td>\n",
       "      <td>0.0761</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)</td>\n",
       "      <td>lncRNA;processed_pseudogene</td>\n",
       "      <td>2;1</td>\n",
       "      <td>ENSG00000238009.6;ENSG00000233750.3</td>\n",
       "      <td>-;+</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>275815</td>\n",
       "      <td>277626</td>\n",
       "      <td>2790</td>\n",
       "      <td>2801</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:275815-277626</td>\n",
       "      <td>11</td>\n",
       "      <td>1811</td>\n",
       "      <td>0.236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>U</td>\n",
       "      <td>ENSG00000228463(100.0)(1.78)</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>-</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>527011</td>\n",
       "      <td>528180</td>\n",
       "      <td>4884</td>\n",
       "      <td>4906</td>\n",
       "      <td>colon_cancer</td>\n",
       "      <td>chr1:527011-528180</td>\n",
       "      <td>22</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>U</td>\n",
       "      <td>NA(0)(0)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  startCpG  endCpG        target              region  \\\n",
       "0  chr1   92977   93393      1045    1051  colon_cancer    chr1:92977-93393   \n",
       "1  chr1   99252   99720      1094    1101  colon_cancer    chr1:99252-99720   \n",
       "2  chr1  132975  133181      1336    1344  colon_cancer  chr1:132975-133181   \n",
       "3  chr1  275815  277626      2790    2801  colon_cancer  chr1:275815-277626   \n",
       "4  chr1  527011  528180      4884    4906  colon_cancer  chr1:527011-528180   \n",
       "\n",
       "  lenCpG    bp  tg_mean  ...  delta_maxmin   ttest  direction  \\\n",
       "0      6   416    0.311  ...         0.330  0.1260          U   \n",
       "1      7   468    0.260  ...         0.314  0.1460          U   \n",
       "2      8   206    0.633  ...         0.344  0.0761          U   \n",
       "3     11  1811    0.236  ...         0.375  0.1640          U   \n",
       "4     22  1169    0.435  ...         0.375  0.0927          U   \n",
       "\n",
       "                                          gene_name  \\\n",
       "0                      ENSG00000238009(100.0)(0.94)   \n",
       "1                      ENSG00000238009(100.0)(1.05)   \n",
       "2  ENSG00000238009(100.0)(0.46);CICP27(100.0)(5.41)   \n",
       "3                      ENSG00000228463(100.0)(1.78)   \n",
       "4                                          NA(0)(0)   \n",
       "\n",
       "                          gene_type gene_level  \\\n",
       "0                            lncRNA          2   \n",
       "1                            lncRNA          2   \n",
       "2       lncRNA;processed_pseudogene        2;1   \n",
       "3  transcribed_processed_pseudogene          1   \n",
       "4                                NA        NaN   \n",
       "\n",
       "                               gene_id strand       genic gene_num  \n",
       "0                    ENSG00000238009.6      -  intragenic        1  \n",
       "1                    ENSG00000238009.6      -  intragenic        1  \n",
       "2  ENSG00000238009.6;ENSG00000233750.3    -;+  intragenic        2  \n",
       "3                   ENSG00000228463.10      -  intragenic        1  \n",
       "4                                   NA     NA  intergenic        0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save files\n",
    "\n",
    "df has columns: ['chr', 'start', 'end', 'startCpG', 'endCpG', 'target', 'region',\n",
    "                'lenCpG', 'bp', 'tg_mean', 'bg_mean', 'delta_means', 'delta_quants',\n",
    "                'delta_maxmin', 'ttest', 'direction', 'gene_name', 'gene_type',\n",
    "                'gene_level', 'gene_id', 'strand', 'gene_num', 'genic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated.tsv with all regions (partial annotation/no dups) has been created\n",
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_per_gene.tsv with all regions (dup regions to split genes) has been created\n",
      "/Users/meghansleeper/Desktop/MATH615/data/colon_cancer_markers_annotated_dmrs.tsv with only significant regions (dup regions with annotations) has been created\n"
     ]
    }
   ],
   "source": [
    "df_perGene.to_csv(output_file_per_gene, index=False, header=True, sep=\"\\t\")\n",
    "df.to_csv(output_file, index=False, header=True, sep=\"\\t\")\n",
    "dmr.to_csv(dmr_output_file, index=False, header=True, sep=\"\\t\")\n",
    "\n",
    "print(output_file + \" with all regions (partial annotation/no dups) has been created\")\n",
    "print(output_file_per_gene + \" with all regions (dup regions to split genes) has been created\")\n",
    "print(dmr_output_file + \" with only significant regions (dup regions with annotations) has been created\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Misc. code chunks from dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # display dmr dataframe where gene_num is greater than 1\n",
    "# df.loc[(df['gene_num'] > 1)]\n",
    "\n",
    "# # return rows where feature is exon in df_perGene\n",
    "# df_perGene[df_perGene.feature=='exon'].tail(25)\n",
    "\n",
    "## skip if you want to keep regions with no genes\n",
    "## Transforms the data frame and removes anything that returned NA\n",
    "## Remove all the intervals that do not overlap with genes\n",
    "# df = df[df['genes'] != \"NA(0)\"].reset_index(drop=True)\n",
    "\n",
    "# # display items in list of lists\n",
    "# print(new_rows[0:5])\n",
    "\n",
    "# # listing column names\n",
    "# column_names = list(df.columns.values)\n",
    "# print(column_names)\n",
    "\n",
    "# # exporting top 100 regions with lowest p-values for annotation in browser (USC Genome Browser has gencode V44 annotation)\n",
    "# df.sort_values(by=['ttest'], inplace=True)\n",
    "# dmr_exp = df.loc[(df['ttest']<0.050)].head(100)\n",
    "# dmr_exp = dmr_exp.drop(['startCpG', 'endCpG', 'target', 'bp',\n",
    "#                    'lenCpG', 'region', 'tg_mean', 'bg_mean', \n",
    "#                    'delta_means', 'delta_quants', 'delta_maxmin',\n",
    "#                    'ttest', 'direction'], axis=1).head(100)\n",
    "# dmr_exp.to_csv('top100_colon_markers.tsv', sep=\"\\t\", header=False, index=False) \n",
    "# dmr_exp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions for keller\n",
    "\n",
    "What do the various chromosome labels mean? \n",
    "Are these contigs that are not included in chromosome contructed references?\n",
    "    Does `M` refer to mitrochondrial?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GL000009.2', 'GL000194.1', 'GL000195.1', 'GL000205.2',\n",
       "       'GL000209.2', 'GL000213.1', 'GL000216.2', 'GL000218.1',\n",
       "       'GL000219.1', 'GL000220.1', 'GL000225.1', 'GL000250.2',\n",
       "       'GL000251.2', 'GL000252.2', 'GL000253.2', 'GL000254.2',\n",
       "       'GL000255.2', 'GL000256.2', 'GL000257.2', 'GL000258.2',\n",
       "       'GL339449.2', 'GL383518.1', 'GL383519.1', 'GL383520.2',\n",
       "       'GL383521.1', 'GL383522.1', 'GL383526.1', 'GL383527.1',\n",
       "       'GL383528.1', 'GL383531.1', 'GL383532.1', 'GL383533.1',\n",
       "       'GL383534.2', 'GL383539.1', 'GL383540.1', 'GL383541.1',\n",
       "       'GL383542.1', 'GL383545.1', 'GL383546.1', 'GL383550.2',\n",
       "       'GL383551.1', 'GL383552.1', 'GL383553.2', 'GL383554.1',\n",
       "       'GL383555.2', 'GL383556.1', 'GL383557.1', 'GL383563.3',\n",
       "       'GL383564.2', 'GL383565.1', 'GL383566.1', 'GL383567.1',\n",
       "       'GL383569.1', 'GL383570.1', 'GL383571.1', 'GL383572.1',\n",
       "       'GL383573.1', 'GL383574.1', 'GL383575.2', 'GL383576.1',\n",
       "       'GL383577.2', 'GL383579.2', 'GL383580.2', 'GL383581.2',\n",
       "       'GL383582.2', 'GL383583.2', 'GL582966.2', 'GL877875.1',\n",
       "       'GL877876.1', 'GL949742.1', 'GL949746.1', 'GL949747.2',\n",
       "       'GL949748.2', 'GL949749.2', 'GL949750.2', 'GL949751.2',\n",
       "       'GL949752.1', 'GL949753.2', 'JH159136.1', 'JH159137.1',\n",
       "       'JH159146.1', 'JH159147.1', 'JH159148.1', 'JH636055.2',\n",
       "       'KB021644.2', 'KB663609.1', 'KI270442.1', 'KI270711.1',\n",
       "       'KI270713.1', 'KI270721.1', 'KI270726.1', 'KI270727.1',\n",
       "       'KI270728.1', 'KI270731.1', 'KI270733.1', 'KI270734.1',\n",
       "       'KI270744.1', 'KI270750.1', 'KI270758.1', 'KI270759.1',\n",
       "       'KI270760.1', 'KI270761.1', 'KI270762.1', 'KI270763.1',\n",
       "       'KI270765.1', 'KI270766.1', 'KI270767.1', 'KI270768.1',\n",
       "       'KI270769.1', 'KI270770.1', 'KI270772.1', 'KI270773.1',\n",
       "       'KI270774.1', 'KI270776.1', 'KI270777.1', 'KI270779.1',\n",
       "       'KI270780.1', 'KI270781.1', 'KI270782.1', 'KI270783.1',\n",
       "       'KI270784.1', 'KI270785.1', 'KI270786.1', 'KI270788.1',\n",
       "       'KI270789.1', 'KI270790.1', 'KI270791.1', 'KI270792.1',\n",
       "       'KI270793.1', 'KI270794.1', 'KI270795.1', 'KI270796.1',\n",
       "       'KI270797.1', 'KI270798.1', 'KI270799.1', 'KI270800.1',\n",
       "       'KI270801.1', 'KI270802.1', 'KI270803.1', 'KI270804.1',\n",
       "       'KI270805.1', 'KI270806.1', 'KI270807.1', 'KI270808.1',\n",
       "       'KI270809.1', 'KI270810.1', 'KI270811.1', 'KI270812.1',\n",
       "       'KI270813.1', 'KI270814.1', 'KI270815.1', 'KI270816.1',\n",
       "       'KI270817.1', 'KI270818.1', 'KI270819.1', 'KI270821.1',\n",
       "       'KI270822.1', 'KI270823.1', 'KI270824.1', 'KI270825.1',\n",
       "       'KI270829.1', 'KI270830.1', 'KI270831.1', 'KI270832.1',\n",
       "       'KI270833.1', 'KI270834.1', 'KI270835.1', 'KI270837.1',\n",
       "       'KI270838.1', 'KI270840.1', 'KI270842.1', 'KI270844.1',\n",
       "       'KI270845.1', 'KI270846.1', 'KI270847.1', 'KI270848.1',\n",
       "       'KI270849.1', 'KI270850.1', 'KI270851.1', 'KI270852.1',\n",
       "       'KI270853.1', 'KI270854.1', 'KI270855.1', 'KI270856.1',\n",
       "       'KI270857.1', 'KI270858.1', 'KI270859.1', 'KI270860.1',\n",
       "       'KI270861.1', 'KI270862.1', 'KI270863.1', 'KI270865.1',\n",
       "       'KI270866.1', 'KI270867.1', 'KI270868.1', 'KI270869.1',\n",
       "       'KI270870.1', 'KI270871.1', 'KI270872.1', 'KI270873.1',\n",
       "       'KI270874.1', 'KI270875.1', 'KI270876.1', 'KI270877.1',\n",
       "       'KI270878.1', 'KI270879.1', 'KI270880.1', 'KI270881.1',\n",
       "       'KI270882.1', 'KI270883.1', 'KI270884.1', 'KI270885.1',\n",
       "       'KI270886.1', 'KI270887.1', 'KI270888.1', 'KI270889.1',\n",
       "       'KI270890.1', 'KI270891.1', 'KI270892.1', 'KI270893.1',\n",
       "       'KI270894.1', 'KI270895.1', 'KI270896.1', 'KI270897.1',\n",
       "       'KI270898.1', 'KI270899.1', 'KI270900.1', 'KI270901.1',\n",
       "       'KI270902.1', 'KI270903.1', 'KI270904.1', 'KI270905.1',\n",
       "       'KI270906.1', 'KI270907.1', 'KI270908.1', 'KI270909.1',\n",
       "       'KI270910.1', 'KI270911.1', 'KI270912.1', 'KI270913.1',\n",
       "       'KI270914.1', 'KI270915.1', 'KI270916.1', 'KI270917.1',\n",
       "       'KI270918.1', 'KI270919.1', 'KI270920.1', 'KI270921.1',\n",
       "       'KI270922.1', 'KI270923.1', 'KI270924.1', 'KI270925.1',\n",
       "       'KI270926.1', 'KI270927.1', 'KI270928.1', 'KI270929.1',\n",
       "       'KI270930.1', 'KI270931.1', 'KI270932.1', 'KI270933.1',\n",
       "       'KI270934.1', 'KI270935.1', 'KI270936.1', 'KI270937.1',\n",
       "       'KI270938.1', 'KN196472.1', 'KN196473.1', 'KN196474.1',\n",
       "       'KN196475.1', 'KN196476.1', 'KN196477.1', 'KN196478.1',\n",
       "       'KN196479.1', 'KN196480.1', 'KN196481.1', 'KN196484.1',\n",
       "       'KN196485.1', 'KN196486.1', 'KN196487.1', 'KN538360.1',\n",
       "       'KN538361.1', 'KN538362.1', 'KN538363.1', 'KN538364.1',\n",
       "       'KN538368.1', 'KN538369.1', 'KN538370.1', 'KN538371.1',\n",
       "       'KN538372.1', 'KN538373.1', 'KQ031383.1', 'KQ031384.1',\n",
       "       'KQ031385.1', 'KQ031387.1', 'KQ031388.1', 'KQ031389.1',\n",
       "       'KQ090014.1', 'KQ090015.1', 'KQ090016.1', 'KQ090018.1',\n",
       "       'KQ090021.1', 'KQ090022.1', 'KQ090026.1', 'KQ090027.1',\n",
       "       'KQ090028.1', 'KQ458382.1', 'KQ458383.1', 'KQ458384.1',\n",
       "       'KQ458385.1', 'KQ458386.1', 'KQ458387.1', 'KQ458388.1',\n",
       "       'KQ759759.2', 'KQ759760.1', 'KQ759761.1', 'KQ759762.2',\n",
       "       'KQ983255.1', 'KQ983256.1', 'KQ983257.1', 'KQ983258.1',\n",
       "       'KV575243.1', 'KV575244.1', 'KV575245.1', 'KV575246.1',\n",
       "       'KV575247.1', 'KV575248.1', 'KV575249.1', 'KV575250.1',\n",
       "       'KV575251.1', 'KV575252.1', 'KV575253.1', 'KV575254.1',\n",
       "       'KV575255.1', 'KV575256.1', 'KV575257.1', 'KV575258.1',\n",
       "       'KV575259.1', 'KV575260.1', 'KV766192.1', 'KV766193.1',\n",
       "       'KV766194.1', 'KV766195.1', 'KV766196.1', 'KV766198.1',\n",
       "       'KV880763.1', 'KV880764.1', 'KV880765.1', 'KV880766.1',\n",
       "       'KV880768.1', 'KZ208904.1', 'KZ208905.1', 'KZ208906.1',\n",
       "       'KZ208907.1', 'KZ208908.1', 'KZ208909.1', 'KZ208911.1',\n",
       "       'KZ208912.1', 'KZ208913.1', 'KZ208914.1', 'KZ208915.1',\n",
       "       'KZ208916.1', 'KZ208917.1', 'KZ208919.1', 'KZ208920.1',\n",
       "       'KZ208921.1', 'KZ208922.1', 'KZ208924.1', 'KZ559102.1',\n",
       "       'KZ559103.1', 'KZ559104.1', 'KZ559105.1', 'KZ559106.1',\n",
       "       'KZ559107.1', 'KZ559108.1', 'KZ559109.1', 'KZ559110.1',\n",
       "       'KZ559111.1', 'KZ559112.1', 'KZ559113.1', 'KZ559114.1',\n",
       "       'KZ559115.1', 'KZ559116.1', 'ML143341.1', 'ML143342.1',\n",
       "       'ML143343.1', 'ML143344.1', 'ML143345.1', 'ML143347.1',\n",
       "       'ML143349.1', 'ML143350.1', 'ML143352.1', 'ML143354.1',\n",
       "       'ML143355.1', 'ML143357.1', 'ML143358.1', 'ML143359.1',\n",
       "       'ML143360.1', 'ML143361.1', 'ML143362.1', 'ML143365.1',\n",
       "       'ML143366.1', 'ML143367.1', 'ML143369.1', 'ML143370.1',\n",
       "       'ML143371.1', 'ML143372.1', 'ML143373.1', 'ML143374.1',\n",
       "       'ML143375.1', 'ML143376.1', 'ML143377.1', 'ML143378.1',\n",
       "       'ML143380.1', 'ML143381.1', 'MU273330.1', 'MU273331.1',\n",
       "       'MU273332.1', 'MU273333.1', 'MU273334.1', 'MU273335.1',\n",
       "       'MU273336.1', 'MU273337.1', 'MU273338.1', 'MU273339.1',\n",
       "       'MU273340.1', 'MU273341.1', 'MU273342.1', 'MU273343.1',\n",
       "       'MU273344.1', 'MU273345.1', 'MU273346.1', 'MU273347.1',\n",
       "       'MU273348.1', 'MU273349.1', 'MU273351.1', 'MU273352.1',\n",
       "       'MU273353.1', 'MU273355.1', 'MU273356.1', 'MU273357.1',\n",
       "       'MU273358.1', 'MU273359.1', 'MU273360.1', 'MU273361.1',\n",
       "       'MU273362.1', 'MU273363.1', 'MU273364.1', 'MU273365.1',\n",
       "       'MU273366.1', 'MU273367.1', 'MU273368.1', 'MU273369.1',\n",
       "       'MU273370.1', 'MU273371.1', 'MU273372.1', 'MU273373.1',\n",
       "       'MU273374.1', 'MU273375.1', 'MU273376.1', 'MU273377.1',\n",
       "       'MU273378.1', 'MU273379.1', 'MU273380.1', 'MU273381.1',\n",
       "       'MU273382.1', 'MU273383.1', 'MU273384.1', 'MU273385.1',\n",
       "       'MU273387.1', 'MU273388.1', 'MU273389.1', 'MU273390.1',\n",
       "       'MU273391.1', 'MU273392.1', 'MU273393.1', 'MU273394.1',\n",
       "       'MU273395.1', 'MU273396.1', 'MU273397.1', 'MU273398.1', 'chr1',\n",
       "       'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16',\n",
       "       'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21', 'chr22',\n",
       "       'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrM',\n",
       "       'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# figure out how to interpret the chromosome labels\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=\n",
    "annot_df.chr.unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "annot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
